{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "61e8049e",
   "metadata": {},
   "source": [
    "# Starting with quality control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8f95ed1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/media/dftortosa/Windows/Users/dftor/Documents/diego_docs/science/other_projects/australian_army_bishop/quality_control'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "os.chdir(\"/media/dftortosa/Windows/Users/dftor/Documents/diego_docs/science/other_projects/australian_army_bishop/quality_control\")\n",
    "\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60260480",
   "metadata": {},
   "source": [
    "## Data available"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22ddd983",
   "metadata": {},
   "source": [
    "I have received a Illumina report with 3 files ([link](https://www.biostars.org/p/51928/)):\n",
    "\n",
    "- The \"FinalReport.txt\" for Illumina raw genotype data generated from Genome Bead Studio for 2.5M (GSGT Version\t2.0.4). This includes a hader with number of SNPs, samples.... and then the data with sample index, sample names, alleles... the first row includes the column names. This is a BPM file.\n",
    "    - From this file, we can obtain the a lgen file. It is just taking the first few columns of the FinalReport. Plink has an option to read the .lgen format and convert it to PED file (see below; [link](https://www.biostars.org/p/13302/))\n",
    "\n",
    "- A SNP map file with the physical positions of the snps.\n",
    "\n",
    "- A sample map file with info for each individual, like the sex, ID.."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72622ad1",
   "metadata": {},
   "source": [
    "It is usually recommended to prepare this data to create a ped file with Plink, which is a tool to process genetic data ([link](https://www.cog-genomics.org/plink/)), perform some filtering and QC analyses and then export as VCF ([link](https://www.biostars.org/p/210516/), [link](https://www.biostars.org/p/135156/), [link](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3066182/)) and use packages like Hail in python.\n",
    "\n",
    "There is an interesting alternative, gtc2vcf ([link](https://github.com/freeseek/gtc2vcf), [link](https://software.broadinstitute.org/software/gtc2vcf/)), which can directly transform Ilumina reports into VCF files from command line. We are going to use Plink, though, because it is much more widely used and there are tutorials and best practice papers using it.\n",
    "\n",
    "In particular, we are going to use the a paper about QC by Ritchie.There is a first version 2011 ([link](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3066182/)) and a second version in 2022 ([link](https://currentprotocols.onlinelibrary.wiley.com/doi/10.1002/cpz1.603)).\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d77ef775",
   "metadata": {},
   "source": [
    "## Plink Installation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec5ba838",
   "metadata": {},
   "source": [
    "I have downloaded Plink ([link](https://www.cog-genomics.org/plink/)) and copied the executable (\"plink\") to `bin`, so we can use Plink from any place just typing `plink`. We are using Plink version 1.9 (Dec 2022)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a8cf4d80",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b6.27 64-bit (10 Dec 2022)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "plink --version"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "713fcf8c",
   "metadata": {},
   "source": [
    "Note that there is Plink 1.9 ([link](https://www.cog-genomics.org/plink/1.9/)) and Plink 2.0 ([link](https://www.cog-genomics.org/plink/2.0/)), these are not connected but different programs. \n",
    "\n",
    "This [threat](https://www.biostars.org/p/299855/#:~:text=The%20main%20difference%20is%20that,for%20a%20while%20to%20come.) of biostars explains the differences:\n",
    "\n",
    "\"The main difference is that plink 1.9 is essentially finished, while plink 2.0 is an alpha-stage program which will have significant unfinished components for a while to come. As a consequence, current development priorities for plink 2.0 are centered around things which are impossible to do with plink 1.9, such as handling multiallelic/phased variants and dosage data and reliably tracking REF/ALT alleles; while things that plink 1.9 already handles perfectly well, such as working with .ped/.map file pairs, have been deliberately deprioritized for now.\n",
    "\n",
    "So, **you should stick to 1.9 as long as it's good enough for the jobs you need to perform. But once you need to do something outside 1.9's scope, you're likely to find that 2.0 already has the additional feature you need** (or it'll be added quickly after you ask for it)\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ff03453",
   "metadata": {},
   "source": [
    "## Plink dummy example"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3d43ea8",
   "metadata": {},
   "source": [
    "Follow the general usage page ([link](https://www.cog-genomics.org/plink/1.9/general_usage)) of plink and run the toy example found in the downloaded folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fbcdb12c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b6.27 64-bit (10 Dec 2022)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2022 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to toy_analysis.log.\n",
      "Options in effect:\n",
      "  --file toy\n",
      "  --freq\n",
      "  --out toy_analysis\n",
      "\n",
      "64162 MB RAM detected; reserving 32081 MB for main workspace.\n",
      ".ped scan complete (for binary autoconversion).\n",
      "Performing single-pass .bed write (2 variants, 2 people).\n",
      "--file: toy_analysis-temporary.bed + toy_analysis-temporary.bim +243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394%\n",
      "toy_analysis-temporary.fam written.\n",
      "2 variants loaded from .bim file.\n",
      "2 people (2 males, 0 females) loaded from .fam.\n",
      "2 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 2 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.75.\n",
      "--freq: Allele frequencies (founders only) written to toy_analysis.frq .\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "cd /media/dftortosa/Windows/Users/dftor/Documents/diego_docs/science/software/plink/plink_linux_x86_64_20221210\n",
    "\n",
    "plink --file toy --freq --out toy_analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6df6334d",
   "metadata": {},
   "source": [
    "Explanation of the arguments\n",
    "\n",
    "- \"--file toy\" tells PLINK to use the genomic data in the text files toy.ped and toy.map. You'll see several other ways to specify input data on the next page.\n",
    "    - toy.map has two rows, one for each SNP ([link](https://www.cog-genomics.org/plink/1.9/formats#map)).\n",
    "        - Chromosome code. PLINK 1.9 also permits contig names here, but most older programs do not.\n",
    "        - Variant identifier\n",
    "        - Position in morgans or centimorgans (optional; also safe to use dummy value of '0')\n",
    "        - Base-pair coordinate \n",
    "        - All lines must have the same number of columns (so either no lines contain the morgans/centimorgans column, or all of them do).\n",
    "    - toy.ped. Contains no header line, and one line per sample with 2V+6 fields where V is the number of variants. \n",
    "        - The first six fields are the same as those in a .fam file:\n",
    "            - Family ID ('FID')\n",
    "            - Within-family ID ('IID'; cannot be '0')\n",
    "            - Within-family ID of father ('0' if father isn't in dataset)\n",
    "            - Within-family ID of mother ('0' if mother isn't in dataset)\n",
    "            - Sex code ('1' = male, '2' = female, '0' = unknown)\n",
    "            - Phenotype value ('1' = control, '2' = case, '-9'/'0'/non-numeric = missing data if case/control)\n",
    "        - The seventh and eighth fields are allele calls for the first variant in the .map file ('0' = no call); the 9th and 10th are allele calls for the second variant; and so on.\n",
    "        - If all alleles are single-character, PLINK 1.9 will correctly parse the more compact \"compound genotype\" variant of this format, where each genotype call is represented as a single two-character string. This does not require the use of an additional loading flag. You can produce such a file with \"--recode compound-genotypes\". \n",
    "        - It is also possible to load .ped files missing some initial fields.\n",
    "- --freq tells PLINK to generate an allele frequency report. The full range of supported calculations is summarized under \"Main functions\" in the sidebar, and the formats of all reports are described in the file formats appendix.\n",
    "- --out is for the output file prefix.\n",
    "\n",
    "So this particular combination makes PLINK calculate allele frequencies in toy.ped + toy.map, and write a report to toy_analysis.frq."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75182d79",
   "metadata": {},
   "source": [
    "Here ([link](https://www.cog-genomics.org/plink/1.9/general_usage)) you can find information about the flag usage in plink. As an example, we are going to work with --out flag.\n",
    "\n",
    "By default, the output files generated by PLINK all have names of the form 'plink.<one of these extensions>'. This is fine for a single run, but as soon as you make more use of PLINK, you'll start causing results from previous runs to be overwritten.\n",
    "\n",
    "Therefore, you usually want to choose a different output file prefix for each run. --out causes 'plink' to be replaced with the prefix you provide. E.g. in the example above, \"--out toy_analysis\" caused PLINK to create a file named toy_analysis.frq instead of plink.frq.\n",
    "\n",
    "Since the prefix is a required parameter, invoking --out without it will cause PLINK to quit during command line parsing:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3f7d262f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: Missing --out parameter.\n",
      "For more information, try \"plink --help <flag name>\" or \"plink --help | more\".\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b6.27 64-bit (10 Dec 2022)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2022 Shaun Purcell, Christopher Chang   GNU General Public License v3\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command 'b'\\nplink --file toy --freq --out\\n'' returned non-zero exit status 5.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mget_ipython\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_cell_magic\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mbash\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mplink --file toy --freq --out\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py:2417\u001b[0m, in \u001b[0;36mInteractiveShell.run_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2415\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[1;32m   2416\u001b[0m     args \u001b[38;5;241m=\u001b[39m (magic_arg_s, cell)\n\u001b[0;32m-> 2417\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2418\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/IPython/core/magics/script.py:153\u001b[0m, in \u001b[0;36mScriptMagics._make_script_magic.<locals>.named_script_magic\u001b[0;34m(line, cell)\u001b[0m\n\u001b[1;32m    151\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    152\u001b[0m     line \u001b[38;5;241m=\u001b[39m script\n\u001b[0;32m--> 153\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshebang\u001b[49m\u001b[43m(\u001b[49m\u001b[43mline\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcell\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.9/site-packages/IPython/core/magics/script.py:305\u001b[0m, in \u001b[0;36mScriptMagics.shebang\u001b[0;34m(self, line, cell)\u001b[0m\n\u001b[1;32m    300\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m args\u001b[38;5;241m.\u001b[39mraise_error \u001b[38;5;129;01mand\u001b[39;00m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    301\u001b[0m     \u001b[38;5;66;03m# If we get here and p.returncode is still None, we must have\u001b[39;00m\n\u001b[1;32m    302\u001b[0m     \u001b[38;5;66;03m# killed it but not yet seen its return code. We don't wait for it,\u001b[39;00m\n\u001b[1;32m    303\u001b[0m     \u001b[38;5;66;03m# in case it's stuck in uninterruptible sleep. -9 = SIGKILL\u001b[39;00m\n\u001b[1;32m    304\u001b[0m     rc \u001b[38;5;241m=\u001b[39m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m9\u001b[39m\n\u001b[0;32m--> 305\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CalledProcessError(rc, cell)\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command 'b'\\nplink --file toy --freq --out\\n'' returned non-zero exit status 5."
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "plink --file toy --freq --out"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc9bc195",
   "metadata": {},
   "source": [
    "##  Exploration of a final report"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23bf40d2",
   "metadata": {},
   "source": [
    "Explore the final report file of one sample."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32ee52fd",
   "metadata": {},
   "source": [
    "We are going to avoid the header, so we skip the first 10 rows leaving a total of 40 rows, i.e., 39 SNPs plus the columns names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c9837c00",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Index</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>SNP Index</th>\n",
       "      <th>SNP Name</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Position</th>\n",
       "      <th>GT Score</th>\n",
       "      <th>GC Score</th>\n",
       "      <th>Allele1 - AB</th>\n",
       "      <th>...</th>\n",
       "      <th>Y Raw</th>\n",
       "      <th>X</th>\n",
       "      <th>Y</th>\n",
       "      <th>B Allele Freq</th>\n",
       "      <th>Log R Ratio</th>\n",
       "      <th>SNP Aux</th>\n",
       "      <th>SNP</th>\n",
       "      <th>ILMN Strand</th>\n",
       "      <th>Top Genomic Sequence</th>\n",
       "      <th>Customer Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1:103380393</td>\n",
       "      <td>1</td>\n",
       "      <td>102914837</td>\n",
       "      <td>0.7987</td>\n",
       "      <td>0.8136</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>958</td>\n",
       "      <td>0.324</td>\n",
       "      <td>0.360</td>\n",
       "      <td>0.4849</td>\n",
       "      <td>0.3097</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1:109439680</td>\n",
       "      <td>1</td>\n",
       "      <td>108897058</td>\n",
       "      <td>0.8792</td>\n",
       "      <td>0.4803</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>131</td>\n",
       "      <td>0.779</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2660</td>\n",
       "      <td>0</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>1:110198788</td>\n",
       "      <td>1</td>\n",
       "      <td>109656166</td>\n",
       "      <td>0.8780</td>\n",
       "      <td>0.9154</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>198</td>\n",
       "      <td>1.170</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0312</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>1:110201112</td>\n",
       "      <td>1</td>\n",
       "      <td>109658490</td>\n",
       "      <td>0.8584</td>\n",
       "      <td>0.4545</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>728</td>\n",
       "      <td>1.154</td>\n",
       "      <td>0.072</td>\n",
       "      <td>0.0188</td>\n",
       "      <td>-0.0746</td>\n",
       "      <td>0</td>\n",
       "      <td>[G/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>1:110201667</td>\n",
       "      <td>1</td>\n",
       "      <td>109659045</td>\n",
       "      <td>0.7815</td>\n",
       "      <td>0.7185</td>\n",
       "      <td>B</td>\n",
       "      <td>...</td>\n",
       "      <td>3269</td>\n",
       "      <td>0.120</td>\n",
       "      <td>1.291</td>\n",
       "      <td>0.9797</td>\n",
       "      <td>-0.0749</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654022</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>654023</td>\n",
       "      <td>Y:8804321-A-C</td>\n",
       "      <td>Y</td>\n",
       "      <td>8936280</td>\n",
       "      <td>0.7229</td>\n",
       "      <td>0.4350</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>157</td>\n",
       "      <td>0.403</td>\n",
       "      <td>0.039</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-1.2565</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/G]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654023</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>654024</td>\n",
       "      <td>Y:8852748-A-C</td>\n",
       "      <td>Y</td>\n",
       "      <td>8984707</td>\n",
       "      <td>0.8575</td>\n",
       "      <td>0.4533</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>147</td>\n",
       "      <td>0.730</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.3919</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/G]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654024</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>654025</td>\n",
       "      <td>Y:8894070-A-T</td>\n",
       "      <td>Y</td>\n",
       "      <td>9026029</td>\n",
       "      <td>0.8175</td>\n",
       "      <td>0.4061</td>\n",
       "      <td>B</td>\n",
       "      <td>...</td>\n",
       "      <td>3099</td>\n",
       "      <td>0.017</td>\n",
       "      <td>1.436</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0729</td>\n",
       "      <td>0</td>\n",
       "      <td>[A/T]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654025</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>654026</td>\n",
       "      <td>Y:9145267-T-C</td>\n",
       "      <td>Y</td>\n",
       "      <td>9307658</td>\n",
       "      <td>0.7893</td>\n",
       "      <td>0.2966</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>201</td>\n",
       "      <td>0.334</td>\n",
       "      <td>0.059</td>\n",
       "      <td>0.0178</td>\n",
       "      <td>-1.3112</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654026</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>654027</td>\n",
       "      <td>Y:9785736-A-G</td>\n",
       "      <td>Y</td>\n",
       "      <td>9948127</td>\n",
       "      <td>0.8452</td>\n",
       "      <td>0.2421</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>124</td>\n",
       "      <td>0.260</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-1.7377</td>\n",
       "      <td>0</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>654027 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample Index Sample ID  Sample Name  SNP Index       SNP Name Chr  \\\n",
       "0                  1  7684BSAO          NaN          1    1:103380393   1   \n",
       "1                  1  7684BSAO          NaN          2    1:109439680   1   \n",
       "2                  1  7684BSAO          NaN          3    1:110198788   1   \n",
       "3                  1  7684BSAO          NaN          4    1:110201112   1   \n",
       "4                  1  7684BSAO          NaN          5    1:110201667   1   \n",
       "...              ...       ...          ...        ...            ...  ..   \n",
       "654022             1  7684BSAO          NaN     654023  Y:8804321-A-C   Y   \n",
       "654023             1  7684BSAO          NaN     654024  Y:8852748-A-C   Y   \n",
       "654024             1  7684BSAO          NaN     654025  Y:8894070-A-T   Y   \n",
       "654025             1  7684BSAO          NaN     654026  Y:9145267-T-C   Y   \n",
       "654026             1  7684BSAO          NaN     654027  Y:9785736-A-G   Y   \n",
       "\n",
       "         Position  GT Score  GC Score Allele1 - AB  ... Y Raw      X      Y  \\\n",
       "0       102914837    0.7987    0.8136            A  ...   958  0.324  0.360   \n",
       "1       108897058    0.8792    0.4803            A  ...   131  0.779  0.019   \n",
       "2       109656166    0.8780    0.9154            A  ...   198  1.170  0.028   \n",
       "3       109658490    0.8584    0.4545            A  ...   728  1.154  0.072   \n",
       "4       109659045    0.7815    0.7185            B  ...  3269  0.120  1.291   \n",
       "...           ...       ...       ...          ...  ...   ...    ...    ...   \n",
       "654022    8936280    0.7229    0.4350            A  ...   157  0.403  0.039   \n",
       "654023    8984707    0.8575    0.4533            A  ...   147  0.730  0.028   \n",
       "654024    9026029    0.8175    0.4061            B  ...  3099  0.017  1.436   \n",
       "654025    9307658    0.7893    0.2966            A  ...   201  0.334  0.059   \n",
       "654026    9948127    0.8452    0.2421            A  ...   124  0.260  0.029   \n",
       "\n",
       "       B Allele Freq Log R Ratio SNP Aux    SNP  ILMN Strand  \\\n",
       "0             0.4849      0.3097       0  [T/C]          BOT   \n",
       "1             0.0000      0.2660       0  [A/G]          TOP   \n",
       "2             0.0000      0.0312       0  [T/C]          BOT   \n",
       "3             0.0188     -0.0746       0  [G/C]          BOT   \n",
       "4             0.9797     -0.0749       0  [T/C]          BOT   \n",
       "...              ...         ...     ...    ...          ...   \n",
       "654022        0.0000     -1.2565       0  [T/G]          BOT   \n",
       "654023        0.0000     -0.3919       0  [T/G]          BOT   \n",
       "654024        1.0000      0.0729       0  [A/T]          TOP   \n",
       "654025        0.0178     -1.3112       0  [T/C]          BOT   \n",
       "654026        0.0000     -1.7377       0  [A/G]          TOP   \n",
       "\n",
       "        Top Genomic Sequence  Customer Strand  \n",
       "0                        NaN              TOP  \n",
       "1                        NaN              TOP  \n",
       "2                        NaN              BOT  \n",
       "3                        NaN              TOP  \n",
       "4                        NaN              BOT  \n",
       "...                      ...              ...  \n",
       "654022                   NaN              TOP  \n",
       "654023                   NaN              TOP  \n",
       "654024                   NaN              BOT  \n",
       "654025                   NaN              BOT  \n",
       "654026                   NaN              TOP  \n",
       "\n",
       "[654027 rows x 30 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "final_report1 = pd.read_csv(\"data/example_data/ILGSA24-17303_FinalReport1.txt\",\n",
    "                           skiprows=10, #skip rows with the header\n",
    "                           delimiter=\"\\t\",\n",
    "                           low_memory=False) \n",
    "    #low_memory: Internally process the file in chunks, \n",
    "    #resulting in lower memory use while parsing, \n",
    "    #but possibly mixed type inference. To ensure no mixed \n",
    "    #types either set False, or specify the type with the dtype parameter. \n",
    "final_report1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55508ef1",
   "metadata": {},
   "source": [
    "See columns names (see this [tutorial](https://www.youtube.com/watch?v=_-Gu9hO8aFM&ab_channel=GenomicsBootCamp)):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ad24e087",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sample Index', 'Sample ID', 'Sample Name', 'SNP Index', 'SNP Name',\n",
       "       'Chr', 'Position', 'GT Score', 'GC Score', 'Allele1 - AB',\n",
       "       'Allele2 - AB', 'Allele1 - Top', 'Allele2 - Top', 'Allele1 - Forward',\n",
       "       'Allele2 - Forward', 'Allele1 - Design', 'Allele2 - Design', 'Theta',\n",
       "       'R', 'X Raw', 'Y Raw', 'X', 'Y', 'B Allele Freq', 'Log R Ratio',\n",
       "       'SNP Aux', 'SNP', 'ILMN Strand', 'Top Genomic Sequence',\n",
       "       'Customer Strand'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_report1.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1045fc5d",
   "metadata": {},
   "source": [
    "**Sample identification**\n",
    "\n",
    "The first three columns are about the sample. At least in our case, we have 1 final report for each sample, i.e., each individual (see below), so these columns should have only one value in the entire file. The same sample index and ID. We do not have sample name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5d0fc043",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique cases are 1? True\n",
      "Unique cases are 1? True\n",
      "Unique cases are 1? True\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "print(f'Unique cases are 1? {len(np.unique(final_report1[\"Sample Index\"]))==1}')\n",
    "print(f'Unique cases are 1? {len(np.unique(final_report1[\"Sample ID\"]))==1}')\n",
    "print(f'Unique cases are 1? {len(np.unique(final_report1[\"Sample Name\"]))==1}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f72ab9d",
   "metadata": {},
   "source": [
    "**SNPs identification**\n",
    "\n",
    "Then, we have the SNP indexes and names. We should have a different value for each row, i.e., the number of unique cases should be the same than the number of rows, i.e., no duplicates. Therefore, **we have as many rows as SNPs**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "0aeb1b15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique cases are the number of rows? True\n",
      "Unique cases are the number of rows? True\n"
     ]
    }
   ],
   "source": [
    "print(f'Unique cases are the number of rows? {len(np.unique(final_report1[\"SNP Index\"]))==final_report1.shape[0]}')\n",
    "print(f'Unique cases are the number of rows? {len(np.unique(final_report1[\"SNP Name\"]))==final_report1.shape[0]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c75be0a",
   "metadata": {},
   "source": [
    "**Chromosome data**\n",
    "\n",
    "Then, we have the chromosome. We have some extrange cases like chromosome 0, XY and MT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "81590d5f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['0', '1', '10', '11', '12', '13', '14', '15', '16', '17', '18',\n",
       "       '19', '2', '20', '21', '22', '3', '4', '5', '6', '7', '8', '9',\n",
       "       'MT', 'X', 'XY', 'Y'], dtype=object)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(final_report1[\"Chr\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa89bb11",
   "metadata": {},
   "source": [
    "We have 210 SNPs out 600K that have chromosome zero, and they also have position zero. These seem to be outdated SNPs as they do not have any physical position."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "cbd82e35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'SNPs with chromosome zero are less than 1000? True'"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'SNPs with chromosome zero are less than 1000? {final_report1.loc[final_report1[\"Chr\"] == \"0\"].shape[0] < 1000}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "59dc98c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Chromosome zero SNPs have also position zero? [ True]'"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'Chromosome zero SNPs have also position zero? {np.unique(final_report1.loc[final_report1[\"Chr\"] == \"0\", \"Position\"]) == 0}'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4517a8b",
   "metadata": {},
   "source": [
    "We have also SNPs in chromosome XY."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "1193fa24",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Index</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>SNP Index</th>\n",
       "      <th>SNP Name</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Position</th>\n",
       "      <th>GT Score</th>\n",
       "      <th>GC Score</th>\n",
       "      <th>Allele1 - AB</th>\n",
       "      <th>...</th>\n",
       "      <th>Y Raw</th>\n",
       "      <th>X</th>\n",
       "      <th>Y</th>\n",
       "      <th>B Allele Freq</th>\n",
       "      <th>Log R Ratio</th>\n",
       "      <th>SNP Aux</th>\n",
       "      <th>SNP</th>\n",
       "      <th>ILMN Strand</th>\n",
       "      <th>Top Genomic Sequence</th>\n",
       "      <th>Customer Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14065</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14066</td>\n",
       "      <td>exm2262791</td>\n",
       "      <td>XY</td>\n",
       "      <td>2675970</td>\n",
       "      <td>0.8615</td>\n",
       "      <td>0.8977</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>192</td>\n",
       "      <td>1.410</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1230</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/G]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16165</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16166</td>\n",
       "      <td>exm2273223</td>\n",
       "      <td>XY</td>\n",
       "      <td>265112</td>\n",
       "      <td>0.7593</td>\n",
       "      <td>0.7487</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>279</td>\n",
       "      <td>1.814</td>\n",
       "      <td>0.055</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2450</td>\n",
       "      <td>0</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16166</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16167</td>\n",
       "      <td>exm2273224</td>\n",
       "      <td>XY</td>\n",
       "      <td>1310262</td>\n",
       "      <td>0.8552</td>\n",
       "      <td>0.8905</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>204</td>\n",
       "      <td>1.235</td>\n",
       "      <td>0.040</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1650</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312826</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>312827</td>\n",
       "      <td>JHU_X.1211573</td>\n",
       "      <td>XY</td>\n",
       "      <td>1111421</td>\n",
       "      <td>0.7731</td>\n",
       "      <td>0.7040</td>\n",
       "      <td>B</td>\n",
       "      <td>...</td>\n",
       "      <td>4575</td>\n",
       "      <td>0.083</td>\n",
       "      <td>1.829</td>\n",
       "      <td>0.9857</td>\n",
       "      <td>-0.1434</td>\n",
       "      <td>0</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312832</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>312833</td>\n",
       "      <td>JHU_X.1215195</td>\n",
       "      <td>XY</td>\n",
       "      <td>1115043</td>\n",
       "      <td>0.7880</td>\n",
       "      <td>0.7967</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>137</td>\n",
       "      <td>0.608</td>\n",
       "      <td>0.027</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.3454</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647731</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>647732</td>\n",
       "      <td>rs9786294</td>\n",
       "      <td>XY</td>\n",
       "      <td>92802202</td>\n",
       "      <td>0.7648</td>\n",
       "      <td>0.6893</td>\n",
       "      <td>B</td>\n",
       "      <td>...</td>\n",
       "      <td>3295</td>\n",
       "      <td>0.114</td>\n",
       "      <td>1.325</td>\n",
       "      <td>0.9915</td>\n",
       "      <td>0.0394</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/G]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647740</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>647741</td>\n",
       "      <td>rs9786334</td>\n",
       "      <td>XY</td>\n",
       "      <td>90826461</td>\n",
       "      <td>0.8043</td>\n",
       "      <td>0.7570</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>221</td>\n",
       "      <td>1.006</td>\n",
       "      <td>0.051</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>-0.3718</td>\n",
       "      <td>0</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647776</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>647777</td>\n",
       "      <td>rs9786468</td>\n",
       "      <td>XY</td>\n",
       "      <td>91577410</td>\n",
       "      <td>0.8548</td>\n",
       "      <td>0.8337</td>\n",
       "      <td>B</td>\n",
       "      <td>...</td>\n",
       "      <td>1530</td>\n",
       "      <td>0.083</td>\n",
       "      <td>0.606</td>\n",
       "      <td>0.9670</td>\n",
       "      <td>-0.0696</td>\n",
       "      <td>0</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648034</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>648035</td>\n",
       "      <td>rs9792836</td>\n",
       "      <td>XY</td>\n",
       "      <td>92590088</td>\n",
       "      <td>0.7972</td>\n",
       "      <td>0.3833</td>\n",
       "      <td>B</td>\n",
       "      <td>...</td>\n",
       "      <td>3397</td>\n",
       "      <td>0.083</td>\n",
       "      <td>1.367</td>\n",
       "      <td>0.9906</td>\n",
       "      <td>-0.2638</td>\n",
       "      <td>0</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>653171</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>653172</td>\n",
       "      <td>w23000898302</td>\n",
       "      <td>XY</td>\n",
       "      <td>937567</td>\n",
       "      <td>0.7552</td>\n",
       "      <td>0.7414</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>266</td>\n",
       "      <td>0.730</td>\n",
       "      <td>0.076</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0512</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/G]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>878 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample Index Sample ID  Sample Name  SNP Index       SNP Name Chr  \\\n",
       "14065              1  7684BSAO          NaN      14066     exm2262791  XY   \n",
       "16165              1  7684BSAO          NaN      16166     exm2273223  XY   \n",
       "16166              1  7684BSAO          NaN      16167     exm2273224  XY   \n",
       "312826             1  7684BSAO          NaN     312827  JHU_X.1211573  XY   \n",
       "312832             1  7684BSAO          NaN     312833  JHU_X.1215195  XY   \n",
       "...              ...       ...          ...        ...            ...  ..   \n",
       "647731             1  7684BSAO          NaN     647732      rs9786294  XY   \n",
       "647740             1  7684BSAO          NaN     647741      rs9786334  XY   \n",
       "647776             1  7684BSAO          NaN     647777      rs9786468  XY   \n",
       "648034             1  7684BSAO          NaN     648035      rs9792836  XY   \n",
       "653171             1  7684BSAO          NaN     653172   w23000898302  XY   \n",
       "\n",
       "        Position  GT Score  GC Score Allele1 - AB  ... Y Raw      X      Y  \\\n",
       "14065    2675970    0.8615    0.8977            A  ...   192  1.410  0.028   \n",
       "16165     265112    0.7593    0.7487            A  ...   279  1.814  0.055   \n",
       "16166    1310262    0.8552    0.8905            A  ...   204  1.235  0.040   \n",
       "312826   1111421    0.7731    0.7040            B  ...  4575  0.083  1.829   \n",
       "312832   1115043    0.7880    0.7967            A  ...   137  0.608  0.027   \n",
       "...          ...       ...       ...          ...  ...   ...    ...    ...   \n",
       "647731  92802202    0.7648    0.6893            B  ...  3295  0.114  1.325   \n",
       "647740  90826461    0.8043    0.7570            A  ...   221  1.006  0.051   \n",
       "647776  91577410    0.8548    0.8337            B  ...  1530  0.083  0.606   \n",
       "648034  92590088    0.7972    0.3833            B  ...  3397  0.083  1.367   \n",
       "653171    937567    0.7552    0.7414            A  ...   266  0.730  0.076   \n",
       "\n",
       "       B Allele Freq Log R Ratio SNP Aux    SNP  ILMN Strand  \\\n",
       "14065         0.0000      0.1230       0  [T/G]          BOT   \n",
       "16165         0.0000      0.2450       0  [A/G]          TOP   \n",
       "16166         0.0000      0.1650       0  [T/C]          BOT   \n",
       "312826        0.9857     -0.1434       0  [A/G]          TOP   \n",
       "312832        0.0000     -0.3454       0  [T/C]          BOT   \n",
       "...              ...         ...     ...    ...          ...   \n",
       "647731        0.9915      0.0394       0  [T/G]          BOT   \n",
       "647740        0.0172     -0.3718       0  [A/G]          TOP   \n",
       "647776        0.9670     -0.0696       0  [A/G]          TOP   \n",
       "648034        0.9906     -0.2638       0  [A/G]          TOP   \n",
       "653171        0.0000      0.0512       0  [T/G]          BOT   \n",
       "\n",
       "        Top Genomic Sequence  Customer Strand  \n",
       "14065                    NaN              TOP  \n",
       "16165                    NaN              BOT  \n",
       "16166                    NaN              BOT  \n",
       "312826                   NaN              TOP  \n",
       "312832                   NaN              TOP  \n",
       "...                      ...              ...  \n",
       "647731                   NaN              BOT  \n",
       "647740                   NaN              BOT  \n",
       "647776                   NaN              TOP  \n",
       "648034                   NaN              BOT  \n",
       "653171                   NaN              TOP  \n",
       "\n",
       "[878 rows x 30 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_report1.loc[final_report1[\"Chr\"] == \"XY\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "2a4697ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'The number of XY SNPs is less than 1000: True'"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'The number of XY SNPs is less than 1000: {final_report1.loc[final_report1[\"Chr\"] == \"XY\"].shape[0]<1000}'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c421196",
   "metadata": {},
   "source": [
    "These seems to be pseudoautosomal regions (PAR), i.e., homologous sequences of nucleotides on the X and Y chromosomes. Therefore, this should be ok, the error is to put them in X or XY while the convention is to set them as XY ([link; section Sex-chromosome marker annotation](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5100670/))."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fcb7339",
   "metadata": {},
   "source": [
    "We have also mitochondrial SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "74ba7259",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Index</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>SNP Index</th>\n",
       "      <th>SNP Name</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Position</th>\n",
       "      <th>GT Score</th>\n",
       "      <th>GC Score</th>\n",
       "      <th>Allele1 - AB</th>\n",
       "      <th>...</th>\n",
       "      <th>Y Raw</th>\n",
       "      <th>X</th>\n",
       "      <th>Y</th>\n",
       "      <th>B Allele Freq</th>\n",
       "      <th>Log R Ratio</th>\n",
       "      <th>SNP Aux</th>\n",
       "      <th>SNP</th>\n",
       "      <th>ILMN Strand</th>\n",
       "      <th>Top Genomic Sequence</th>\n",
       "      <th>Customer Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8158</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8159</td>\n",
       "      <td>200610-1</td>\n",
       "      <td>MT</td>\n",
       "      <td>2755</td>\n",
       "      <td>0.6796</td>\n",
       "      <td>0.5358</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>775</td>\n",
       "      <td>3.385</td>\n",
       "      <td>0.219</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0727</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8159</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8160</td>\n",
       "      <td>200610-100</td>\n",
       "      <td>MT</td>\n",
       "      <td>15172</td>\n",
       "      <td>0.6310</td>\n",
       "      <td>0.2317</td>\n",
       "      <td>B</td>\n",
       "      <td>...</td>\n",
       "      <td>4100</td>\n",
       "      <td>0.432</td>\n",
       "      <td>1.586</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>-0.2949</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8160</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8161</td>\n",
       "      <td>200610-102</td>\n",
       "      <td>MT</td>\n",
       "      <td>125</td>\n",
       "      <td>0.7739</td>\n",
       "      <td>0.3583</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>299</td>\n",
       "      <td>1.618</td>\n",
       "      <td>0.071</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.0262</td>\n",
       "      <td>0</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8161</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8162</td>\n",
       "      <td>200610-105</td>\n",
       "      <td>MT</td>\n",
       "      <td>236</td>\n",
       "      <td>0.6705</td>\n",
       "      <td>0.5197</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>371</td>\n",
       "      <td>3.063</td>\n",
       "      <td>0.067</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1016</td>\n",
       "      <td>0</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8162</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8163</td>\n",
       "      <td>200610-110</td>\n",
       "      <td>MT</td>\n",
       "      <td>3027</td>\n",
       "      <td>0.7420</td>\n",
       "      <td>0.6486</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>368</td>\n",
       "      <td>2.516</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1526</td>\n",
       "      <td>0</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545420</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>545421</td>\n",
       "      <td>rs62581340</td>\n",
       "      <td>MT</td>\n",
       "      <td>16356</td>\n",
       "      <td>0.6619</td>\n",
       "      <td>0.5046</td>\n",
       "      <td>B</td>\n",
       "      <td>...</td>\n",
       "      <td>1913</td>\n",
       "      <td>0.178</td>\n",
       "      <td>0.760</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545421</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>545422</td>\n",
       "      <td>rs62581341</td>\n",
       "      <td>MT</td>\n",
       "      <td>16362</td>\n",
       "      <td>0.6257</td>\n",
       "      <td>0.4430</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>407</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.140</td>\n",
       "      <td>0.0804</td>\n",
       "      <td>-1.7221</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>591382</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>591383</td>\n",
       "      <td>rs75214962</td>\n",
       "      <td>MT</td>\n",
       "      <td>11197</td>\n",
       "      <td>0.7111</td>\n",
       "      <td>0.2970</td>\n",
       "      <td>B</td>\n",
       "      <td>...</td>\n",
       "      <td>2214</td>\n",
       "      <td>0.179</td>\n",
       "      <td>0.883</td>\n",
       "      <td>0.9925</td>\n",
       "      <td>-0.3114</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593585</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>593586</td>\n",
       "      <td>rs75577869</td>\n",
       "      <td>MT</td>\n",
       "      <td>13020</td>\n",
       "      <td>0.6882</td>\n",
       "      <td>0.5512</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>400</td>\n",
       "      <td>1.397</td>\n",
       "      <td>0.115</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.1041</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604840</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>604841</td>\n",
       "      <td>rs77417897</td>\n",
       "      <td>MT</td>\n",
       "      <td>13117</td>\n",
       "      <td>0.6551</td>\n",
       "      <td>0.4927</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>674</td>\n",
       "      <td>2.326</td>\n",
       "      <td>0.181</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.0907</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1138 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample Index Sample ID  Sample Name  SNP Index    SNP Name Chr  \\\n",
       "8158               1  7684BSAO          NaN       8159    200610-1  MT   \n",
       "8159               1  7684BSAO          NaN       8160  200610-100  MT   \n",
       "8160               1  7684BSAO          NaN       8161  200610-102  MT   \n",
       "8161               1  7684BSAO          NaN       8162  200610-105  MT   \n",
       "8162               1  7684BSAO          NaN       8163  200610-110  MT   \n",
       "...              ...       ...          ...        ...         ...  ..   \n",
       "545420             1  7684BSAO          NaN     545421  rs62581340  MT   \n",
       "545421             1  7684BSAO          NaN     545422  rs62581341  MT   \n",
       "591382             1  7684BSAO          NaN     591383  rs75214962  MT   \n",
       "593585             1  7684BSAO          NaN     593586  rs75577869  MT   \n",
       "604840             1  7684BSAO          NaN     604841  rs77417897  MT   \n",
       "\n",
       "        Position  GT Score  GC Score Allele1 - AB  ... Y Raw      X      Y  \\\n",
       "8158        2755    0.6796    0.5358            A  ...   775  3.385  0.219   \n",
       "8159       15172    0.6310    0.2317            B  ...  4100  0.432  1.586   \n",
       "8160         125    0.7739    0.3583            A  ...   299  1.618  0.071   \n",
       "8161         236    0.6705    0.5197            A  ...   371  3.063  0.067   \n",
       "8162        3027    0.7420    0.6486            A  ...   368  2.516  0.081   \n",
       "...          ...       ...       ...          ...  ...   ...    ...    ...   \n",
       "545420     16356    0.6619    0.5046            B  ...  1913  0.178  0.760   \n",
       "545421     16362    0.6257    0.4430            A  ...   407  0.464  0.140   \n",
       "591382     11197    0.7111    0.2970            B  ...  2214  0.179  0.883   \n",
       "593585     13020    0.6882    0.5512            A  ...   400  1.397  0.115   \n",
       "604840     13117    0.6551    0.4927            A  ...   674  2.326  0.181   \n",
       "\n",
       "       B Allele Freq Log R Ratio SNP Aux    SNP  ILMN Strand  \\\n",
       "8158          0.0000      0.0727       0  [T/C]          BOT   \n",
       "8159          1.0000     -0.2949       0  [T/C]          BOT   \n",
       "8160          0.0000     -0.0262       0  [A/G]          TOP   \n",
       "8161          0.0000      0.1016       0  [A/G]          TOP   \n",
       "8162          0.0000      0.1526       0  [A/G]          TOP   \n",
       "...              ...         ...     ...    ...          ...   \n",
       "545420        1.0000      0.0146       0  [T/C]          BOT   \n",
       "545421        0.0804     -1.7221       0  [T/C]          BOT   \n",
       "591382        0.9925     -0.3114       0  [T/C]          BOT   \n",
       "593585        0.0000     -0.1041       0  [T/C]          BOT   \n",
       "604840        0.0000     -0.0907       0  [T/C]          BOT   \n",
       "\n",
       "        Top Genomic Sequence  Customer Strand  \n",
       "8158                     NaN              TOP  \n",
       "8159                     NaN              TOP  \n",
       "8160                     NaN              BOT  \n",
       "8161                     NaN              BOT  \n",
       "8162                     NaN              BOT  \n",
       "...                      ...              ...  \n",
       "545420                   NaN              BOT  \n",
       "545421                   NaN              BOT  \n",
       "591382                   NaN              TOP  \n",
       "593585                   NaN              TOP  \n",
       "604840                   NaN              TOP  \n",
       "\n",
       "[1138 rows x 30 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_report1.loc[final_report1[\"Chr\"] == \"MT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "86b83a90",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'The number of mito SNPs is less than 1500: True'"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'The number of mito SNPs is less than 1500: {final_report1.loc[final_report1[\"Chr\"] == \"MT\"].shape[0]<1500}'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43d1f085",
   "metadata": {},
   "source": [
    "**Position**\n",
    "\n",
    "Then, we have the position, which is an integer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8c3acb6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The type of Position is integer? True\n",
      "There are no NAs? True\n"
     ]
    }
   ],
   "source": [
    "print(f'The type of Position is integer? {final_report1[\"Position\"].dtype == \"int64\"}')\n",
    "print(f'There are no NAs? {final_report1.loc[final_report1[\"Position\"].isna()].shape[0] == 0}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f336ae41",
   "metadata": {},
   "source": [
    "**Quality measures**\n",
    "\n",
    "Then, we have the GT and GC scores, which are quality measures."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f86fe6c8",
   "metadata": {},
   "source": [
    "The most important QC parameter is the GenTrain score. The GenTrain score is computed from the GenTrain 2.0 clustering algorithm. It is a measurement of SNP calling quality, ranging from 0 to 1, with higher value meaning better quality ([link](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6171493/))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "bddd6a3d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The type of GT score is integer? True\n",
      "There are no NAs? True\n",
      "Min is equal of higher than 0? True\n",
      "Max is equal or lower than 1? True\n"
     ]
    }
   ],
   "source": [
    "print(f'The type of GT score is integer? {final_report1[\"GT Score\"].dtype == \"float64\"}')\n",
    "print(f'There are no NAs? {final_report1.loc[final_report1[\"GT Score\"].isna()].shape[0] == 0}')\n",
    "print(f'Min is equal of higher than 0? {np.min(final_report1[\"GT Score\"]) >= 0}')\n",
    "print(f'Max is equal or lower than 1? {np.max(final_report1[\"GT Score\"]) <= 1}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "239bfca0",
   "metadata": {},
   "source": [
    "The second most important QC parameter is the cluster separation score, which measures how well the AA, AB and BB clusters are separated. The cluster separation score also ranges from 0 to 1, with higher meaning better (more separation; [link](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6171493/)). \n",
    "\n",
    "In the link cited, in figure 4, you can see several cases for specific snps where the separation between clusters (AA, AB and BB) is not clear (left panels). Meaning that there individuals are not clearly differentiated for this SNP. With realignement they get a better separation and a better score."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bddea4cb",
   "metadata": {},
   "source": [
    "NOT PRESENT IN MY DATASET"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb1f1ea5",
   "metadata": {},
   "source": [
    "The third most important QC parameter is call frequency, which measures the percentage of samples with successful calls for that SNP. The call frequency also ranges from 0 to 1, with higher meaning more samples have successful calls for this SNP ([link](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6171493/))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1f8fe36e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The type of GT score is integer? True\n",
      "There are no NAs? True\n",
      "Min is equal of higher than 0? True\n",
      "Max is equal or lower than 1? True\n"
     ]
    }
   ],
   "source": [
    "print(f'The type of GT score is integer? {final_report1[\"GC Score\"].dtype == \"float64\"}')\n",
    "print(f'There are no NAs? {final_report1.loc[final_report1[\"GC Score\"].isna()].shape[0] == 0}')\n",
    "print(f'Min is equal of higher than 0? {np.min(final_report1[\"GC Score\"]) >= 0}')\n",
    "print(f'Max is equal or lower than 1? {np.max(final_report1[\"GC Score\"]) <= 1}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68a4658e",
   "metadata": {},
   "source": [
    "**Log R ratio and BAF**\n",
    "\n",
    "These metrics could be used to detect copy number variation (e.g., deletions) without having whole genome sequencing ([Nandolo et al 2018](https://www.youtube.com/watch?v=_-Gu9hO8aFM&ab_channel=GenomicsBootCamp))."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "537e7ccc",
   "metadata": {},
   "source": [
    "**Alleles**\n",
    "\n",
    "One of the biggest weakness of the Illumina genotyping array design is Illumina’s definition of strand. As DNA is double-stranded, significant SNPs from GWAS need to be presented with their strand information to properly report the risk alleles. This unfortunately has not been a standard practice. The most intuitive definition of strand is to use the human genome reference as the forward strand. Defying logic, Illumina introduced a more convoluted definition of strand: top and bottom [25], which has caused great confusion with reference to the forward and reverse strand [26, 27] ([link](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6171493/)).\n",
    "\n",
    "DNA strand designations ([link](https://www.biostars.org/p/4885/)):\n",
    "- Customer Strand:Same as the Source strand. For custom content, it is the strand submitted by the customer for probe design.\n",
    "- ILMN Strand, a.k.a. Design Strand: The strand used by Illumina to design probes based on thermodynamic stability and locus specificity according to NCBI BLAST. For this reason, it can differ from the Customer/ Source strand.\n",
    "- Forward/Reverse (Fwd/Rev) Strand: Used by dbSNP, **Fwd/Rev designations can change with NCBI Genome Build updates, so Genome Build must be specified when reporting Fwd/Rev strands**. 1. For SNPs in standard array products, Fwd strand = Source strand, and originates from dbSNP. 2. For custom array product SNPs without rsid’s, the customer can identify the Source strand as Fwd or Rev, based on their own criteria. Illumina custom product files use the customer’s Fwd/Rev designations. Note: The Fwd strand, as identified in Illumina standard product files, should not be confused with Plus (+) strand, which HapMap interchangeably calls the “forward strand.”\n",
    "- Plus/Minus (+/-) Strand: The standard designation for all eukaryotic organisms used by HapMapand 1000 Genomes Project. The 5′ end of the (+) strand is at the tip of the short arm (p arm) of the chromosome and the 5′ end of the (-) strand is at the tip of the long arm (q arm). (+/-) designations can change with NCBI Genome Build updates, so Genome Build must be specified when reporting (+/-) strands.\n",
    "- Source Strand: Same as the Customer strand. The strand submitted to the Illumina designer for probe design. 1. For standard SNPs, it is the Fwd strand as reported in the source database (i.e., dbSNP). 2. Custom content can be reported as rsid’s or as the DNA sequences or chromosomal regions, depending on the format submitted by the customer.\n",
    "- Top/Bottom (Top/Bot) Strand:Top/Bot nomenclature was developed by Illumina using sequence-based context to assign strand designations that does not change regardless of database or genome assembly used. (e.g., depending on the NCBI Genome Build referenced, strand and allele designations can change). Top/Bot is not directly related to Fwd/Rev or (+/-).Top/Bot strand is determined by examining the SNP and the surrounding DNA sequence and it only applies to SNPs with two possible alleles. See the Top/Bot A/B Allele bulletin for more details ([link](https://www.illumina.com/documents/products/technotes/technote_topbot.pdf), [link](https://support.illumina.com/bulletins/2017/06/how-to-interpret-dna-strand-and-allele-information-for-infinium-.html))."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "603597d9",
   "metadata": {},
   "source": [
    "See the different allele names for some SNPs with rs number:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "02961538",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP Name</th>\n",
       "      <th>Allele1 - AB</th>\n",
       "      <th>Allele2 - AB</th>\n",
       "      <th>Allele1 - Top</th>\n",
       "      <th>Allele2 - Top</th>\n",
       "      <th>Allele1 - Forward</th>\n",
       "      <th>Allele2 - Forward</th>\n",
       "      <th>Allele1 - Design</th>\n",
       "      <th>Allele2 - Design</th>\n",
       "      <th>ILMN Strand</th>\n",
       "      <th>SNP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12303</th>\n",
       "      <td>chr10-96521657:rs12248560-TC</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>BOT</td>\n",
       "      <td>[T/C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12304</th>\n",
       "      <td>chr10-96535210:rs72552267</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>TOP</td>\n",
       "      <td>[A/G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12305</th>\n",
       "      <td>chr10-96541616:rs4244285-AG</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>BOT</td>\n",
       "      <td>[T/C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12335</th>\n",
       "      <td>chr16-31107689:rs9923231-CT</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>TOP</td>\n",
       "      <td>[A/G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12369</th>\n",
       "      <td>chr22-42525044:rs1135824</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>BOT</td>\n",
       "      <td>[T/C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652993</th>\n",
       "      <td>rs9999844</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>BOT</td>\n",
       "      <td>[T/C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652994</th>\n",
       "      <td>rs9999847</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>TOP</td>\n",
       "      <td>[A/G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652995</th>\n",
       "      <td>rs999986</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>BOT</td>\n",
       "      <td>[T/G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652996</th>\n",
       "      <td>rs999989</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>BOT</td>\n",
       "      <td>[T/G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652997</th>\n",
       "      <td>rs999994</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>TOP</td>\n",
       "      <td>[A/G]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>629614 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            SNP Name Allele1 - AB Allele2 - AB Allele1 - Top  \\\n",
       "12303   chr10-96521657:rs12248560-TC            B            B             G   \n",
       "12304      chr10-96535210:rs72552267            B            B             G   \n",
       "12305    chr10-96541616:rs4244285-AG            B            B             G   \n",
       "12335    chr16-31107689:rs9923231-CT            A            B             A   \n",
       "12369       chr22-42525044:rs1135824            A            A             A   \n",
       "...                              ...          ...          ...           ...   \n",
       "652993                     rs9999844            B            B             G   \n",
       "652994                     rs9999847            B            B             G   \n",
       "652995                      rs999986            A            B             A   \n",
       "652996                      rs999989            A            A             A   \n",
       "652997                      rs999994            B            B             G   \n",
       "\n",
       "       Allele2 - Top Allele1 - Forward Allele2 - Forward Allele1 - Design  \\\n",
       "12303              G                 C                 C                C   \n",
       "12304              G                 G                 G                G   \n",
       "12305              G                 G                 G                C   \n",
       "12335              G                 T                 C                A   \n",
       "12369              A                 T                 T                T   \n",
       "...              ...               ...               ...              ...   \n",
       "652993             G                 G                 G                C   \n",
       "652994             G                 C                 C                G   \n",
       "652995             C                 T                 G                T   \n",
       "652996             A                 T                 T                T   \n",
       "652997             G                 G                 G                G   \n",
       "\n",
       "       Allele2 - Design ILMN Strand    SNP  \n",
       "12303                 C         BOT  [T/C]  \n",
       "12304                 G         TOP  [A/G]  \n",
       "12305                 C         BOT  [T/C]  \n",
       "12335                 G         TOP  [A/G]  \n",
       "12369                 T         BOT  [T/C]  \n",
       "...                 ...         ...    ...  \n",
       "652993                C         BOT  [T/C]  \n",
       "652994                G         TOP  [A/G]  \n",
       "652995                G         BOT  [T/G]  \n",
       "652996                T         BOT  [T/G]  \n",
       "652997                G         TOP  [A/G]  \n",
       "\n",
       "[629614 rows x 11 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_report1.loc[final_report1[\"SNP Name\"].apply(lambda x: \"rs\" in x), ['SNP Name', \n",
    "                                                                         'Allele1 - AB', \n",
    "                                                                         'Allele2 - AB', \n",
    "                                                                         'Allele1 - Top', \n",
    "                                                                         'Allele2 - Top', \n",
    "                                                                         'Allele1 - Forward',\n",
    "                                                                         'Allele2 - Forward', \n",
    "                                                                         'Allele1 - Design', \n",
    "                                                                         'Allele2 - Design',\n",
    "                                                                         'ILMN Strand', \n",
    "                                                                         'SNP']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdfcf812",
   "metadata": {},
   "source": [
    "For example, rs9999844 has for foward, the aleles indicated in dbSNP. But this does not work always. For example, according to dbSNP, rs999994 is C/T, but we have G.\n",
    "\n",
    "We will use Allele-Foward in general, but we will have to check the alleles at some point, maybe after selecting SNPs for the polygenic score?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68f7473f",
   "metadata": {},
   "source": [
    "**Other columns**\n",
    "\n",
    "There still other columns, probably related to the plots in the report of illumina, but we are not going to explore them for now unless the QC tutorials make use of them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8a75d5ac",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sample Index', 'Sample ID', 'Sample Name', 'SNP Index', 'SNP Name',\n",
       "       'Chr', 'Position', 'GT Score', 'GC Score', 'Allele1 - AB',\n",
       "       'Allele2 - AB', 'Allele1 - Top', 'Allele2 - Top', 'Allele1 - Forward',\n",
       "       'Allele2 - Forward', 'Allele1 - Design', 'Allele2 - Design', 'Theta',\n",
       "       'R', 'X Raw', 'Y Raw', 'X', 'Y', 'B Allele Freq', 'Log R Ratio',\n",
       "       'SNP Aux', 'SNP', 'ILMN Strand', 'Top Genomic Sequence',\n",
       "       'Customer Strand'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_report1.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fa4e34b",
   "metadata": {},
   "source": [
    "## Load and explore the sample and snp maps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09e78255",
   "metadata": {},
   "source": [
    "We need these maps in order to generate the inputs for Plink (see below)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0d0074c",
   "metadata": {},
   "source": [
    "**SNP map**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0ffd8971",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Index</th>\n",
       "      <th>Name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>GenTrain Score</th>\n",
       "      <th>SNP</th>\n",
       "      <th>ILMN Strand</th>\n",
       "      <th>Customer Strand</th>\n",
       "      <th>NormID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1:103380393</td>\n",
       "      <td>1</td>\n",
       "      <td>102914837</td>\n",
       "      <td>0.7987</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>TOP</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1:109439680</td>\n",
       "      <td>1</td>\n",
       "      <td>108897058</td>\n",
       "      <td>0.8792</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>TOP</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1:110198788</td>\n",
       "      <td>1</td>\n",
       "      <td>109656166</td>\n",
       "      <td>0.8780</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>BOT</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>1:110201112</td>\n",
       "      <td>1</td>\n",
       "      <td>109658490</td>\n",
       "      <td>0.8584</td>\n",
       "      <td>[G/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>TOP</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1:110201667</td>\n",
       "      <td>1</td>\n",
       "      <td>109659045</td>\n",
       "      <td>0.7815</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>BOT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654022</th>\n",
       "      <td>654023</td>\n",
       "      <td>Y:8804321-A-C</td>\n",
       "      <td>Y</td>\n",
       "      <td>8936280</td>\n",
       "      <td>0.7229</td>\n",
       "      <td>[T/G]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>TOP</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654023</th>\n",
       "      <td>654024</td>\n",
       "      <td>Y:8852748-A-C</td>\n",
       "      <td>Y</td>\n",
       "      <td>8984707</td>\n",
       "      <td>0.8575</td>\n",
       "      <td>[T/G]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>TOP</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654024</th>\n",
       "      <td>654025</td>\n",
       "      <td>Y:8894070-A-T</td>\n",
       "      <td>Y</td>\n",
       "      <td>9026029</td>\n",
       "      <td>0.8175</td>\n",
       "      <td>[A/T]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>BOT</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654025</th>\n",
       "      <td>654026</td>\n",
       "      <td>Y:9145267-T-C</td>\n",
       "      <td>Y</td>\n",
       "      <td>9307658</td>\n",
       "      <td>0.7893</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>BOT</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654026</th>\n",
       "      <td>654027</td>\n",
       "      <td>Y:9785736-A-G</td>\n",
       "      <td>Y</td>\n",
       "      <td>9948127</td>\n",
       "      <td>0.8452</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>TOP</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>654027 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Index           Name Chromosome   Position  GenTrain Score    SNP  \\\n",
       "0            1    1:103380393          1  102914837          0.7987  [T/C]   \n",
       "1            2    1:109439680          1  108897058          0.8792  [A/G]   \n",
       "2            3    1:110198788          1  109656166          0.8780  [T/C]   \n",
       "3            4    1:110201112          1  109658490          0.8584  [G/C]   \n",
       "4            5    1:110201667          1  109659045          0.7815  [T/C]   \n",
       "...        ...            ...        ...        ...             ...    ...   \n",
       "654022  654023  Y:8804321-A-C          Y    8936280          0.7229  [T/G]   \n",
       "654023  654024  Y:8852748-A-C          Y    8984707          0.8575  [T/G]   \n",
       "654024  654025  Y:8894070-A-T          Y    9026029          0.8175  [A/T]   \n",
       "654025  654026  Y:9145267-T-C          Y    9307658          0.7893  [T/C]   \n",
       "654026  654027  Y:9785736-A-G          Y    9948127          0.8452  [A/G]   \n",
       "\n",
       "       ILMN Strand Customer Strand  NormID  \n",
       "0              BOT             TOP      28  \n",
       "1              TOP             TOP       1  \n",
       "2              BOT             BOT       4  \n",
       "3              BOT             TOP     101  \n",
       "4              BOT             BOT       1  \n",
       "...            ...             ...     ...  \n",
       "654022         BOT             TOP       3  \n",
       "654023         BOT             TOP       3  \n",
       "654024         TOP             BOT     203  \n",
       "654025         BOT             BOT       3  \n",
       "654026         TOP             TOP       3  \n",
       "\n",
       "[654027 rows x 9 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_map = pd.read_csv(\"data/example_data/SNP_Map.txt\",\n",
    "                      delimiter=\"\\t\", \n",
    "                      header=0,\n",
    "                      low_memory=False) \n",
    "    #low_memory: Internally process the file in chunks, \n",
    "    #resulting in lower memory use while parsing, \n",
    "    #but possibly mixed type inference. To ensure no mixed \n",
    "    #types either set False, or specify the type with the dtype parameter. \n",
    "snp_map"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3d7b405",
   "metadata": {},
   "source": [
    "Check we have the same number of SNPs in the final report and the SNP map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e118fe3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Number of SNP matches between SNP map and final report? True'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'Number of SNP matches between SNP map and final report? {snp_map.shape[0] == final_report1.shape[0]}'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "947dbf72",
   "metadata": {},
   "source": [
    "Therefore, we have a row per SNP, having each SNP an Index and a Name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a029af90",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique cases are 1? True\n",
      "Unique cases are 1? True\n"
     ]
    }
   ],
   "source": [
    "print(f'Unique cases are the number of rows? {len(np.unique(snp_map[\"Index\"]))==snp_map.shape[0]}')\n",
    "print(f'Unique cases are the number of rows? {len(np.unique(snp_map[\"Name\"]))==snp_map.shape[0]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ae52a93",
   "metadata": {},
   "source": [
    "Each SNP has its chromose, having autosomals but also MT, zero cases and sex chromosomes.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "4cce34b3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['0', '1', '10', '11', '12', '13', '14', '15', '16', '17', '18',\n",
       "       '19', '2', '20', '21', '22', '3', '4', '5', '6', '7', '8', '9',\n",
       "       'MT', 'X', 'XY', 'Y'], dtype=object)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(snp_map[\"Chromosome\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a8be7d1",
   "metadata": {},
   "source": [
    "We have SNPS that have chromosome zero, and they also have position zero. They seem to be outdated SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "2bff5246",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'SNPs with chromosome zero are less than 1000? True'"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'SNPs with chromosome zero are less than 1000? {snp_map.loc[snp_map[\"Chromosome\"] == \"0\"].shape[0] < 1000}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "798d5334",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Chromosome zero SNPs have also position zero? [ True]'"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'Chromosome zero SNPs have also position zero? {np.unique(snp_map.loc[snp_map[\"Chromosome\"] == \"0\", \"Position\"]) == 0}'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53f36eb7",
   "metadata": {},
   "source": [
    "We have also SNPs in chromosome XY."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "2444ddae",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Index</th>\n",
       "      <th>Name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>GenTrain Score</th>\n",
       "      <th>SNP</th>\n",
       "      <th>ILMN Strand</th>\n",
       "      <th>Customer Strand</th>\n",
       "      <th>NormID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14065</th>\n",
       "      <td>14066</td>\n",
       "      <td>exm2262791</td>\n",
       "      <td>XY</td>\n",
       "      <td>2675970</td>\n",
       "      <td>0.8615</td>\n",
       "      <td>[T/G]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>TOP</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16165</th>\n",
       "      <td>16166</td>\n",
       "      <td>exm2273223</td>\n",
       "      <td>XY</td>\n",
       "      <td>265112</td>\n",
       "      <td>0.7593</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>BOT</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16166</th>\n",
       "      <td>16167</td>\n",
       "      <td>exm2273224</td>\n",
       "      <td>XY</td>\n",
       "      <td>1310262</td>\n",
       "      <td>0.8552</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>BOT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312826</th>\n",
       "      <td>312827</td>\n",
       "      <td>JHU_X.1211573</td>\n",
       "      <td>XY</td>\n",
       "      <td>1111421</td>\n",
       "      <td>0.7731</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>TOP</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312832</th>\n",
       "      <td>312833</td>\n",
       "      <td>JHU_X.1215195</td>\n",
       "      <td>XY</td>\n",
       "      <td>1115043</td>\n",
       "      <td>0.7880</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>TOP</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647731</th>\n",
       "      <td>647732</td>\n",
       "      <td>rs9786294</td>\n",
       "      <td>XY</td>\n",
       "      <td>92802202</td>\n",
       "      <td>0.7648</td>\n",
       "      <td>[T/G]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>BOT</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647740</th>\n",
       "      <td>647741</td>\n",
       "      <td>rs9786334</td>\n",
       "      <td>XY</td>\n",
       "      <td>90826461</td>\n",
       "      <td>0.8043</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>BOT</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647776</th>\n",
       "      <td>647777</td>\n",
       "      <td>rs9786468</td>\n",
       "      <td>XY</td>\n",
       "      <td>91577410</td>\n",
       "      <td>0.8548</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>TOP</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648034</th>\n",
       "      <td>648035</td>\n",
       "      <td>rs9792836</td>\n",
       "      <td>XY</td>\n",
       "      <td>92590088</td>\n",
       "      <td>0.7972</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>BOT</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>653171</th>\n",
       "      <td>653172</td>\n",
       "      <td>w23000898302</td>\n",
       "      <td>XY</td>\n",
       "      <td>937567</td>\n",
       "      <td>0.7552</td>\n",
       "      <td>[T/G]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>TOP</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>878 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Index           Name Chromosome  Position  GenTrain Score    SNP  \\\n",
       "14065    14066     exm2262791         XY   2675970          0.8615  [T/G]   \n",
       "16165    16166     exm2273223         XY    265112          0.7593  [A/G]   \n",
       "16166    16167     exm2273224         XY   1310262          0.8552  [T/C]   \n",
       "312826  312827  JHU_X.1211573         XY   1111421          0.7731  [A/G]   \n",
       "312832  312833  JHU_X.1215195         XY   1115043          0.7880  [T/C]   \n",
       "...        ...            ...        ...       ...             ...    ...   \n",
       "647731  647732      rs9786294         XY  92802202          0.7648  [T/G]   \n",
       "647740  647741      rs9786334         XY  90826461          0.8043  [A/G]   \n",
       "647776  647777      rs9786468         XY  91577410          0.8548  [A/G]   \n",
       "648034  648035      rs9792836         XY  92590088          0.7972  [A/G]   \n",
       "653171  653172   w23000898302         XY    937567          0.7552  [T/G]   \n",
       "\n",
       "       ILMN Strand Customer Strand  NormID  \n",
       "14065          BOT             TOP       6  \n",
       "16165          TOP             BOT       6  \n",
       "16166          BOT             BOT       1  \n",
       "312826         TOP             TOP      34  \n",
       "312832         BOT             TOP      34  \n",
       "...            ...             ...     ...  \n",
       "647731         BOT             BOT       3  \n",
       "647740         TOP             BOT       3  \n",
       "647776         TOP             TOP       3  \n",
       "648034         TOP             BOT       3  \n",
       "653171         BOT             TOP       3  \n",
       "\n",
       "[878 rows x 9 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_map.loc[snp_map[\"Chromosome\"] == \"XY\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "3691de0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'The number of XY SNPs is less than 1000: True'"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'The number of XY SNPs is less than 1000: {snp_map.loc[snp_map[\"Chromosome\"] == \"XY\"].shape[0]<1000}'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61f91077",
   "metadata": {},
   "source": [
    "These seems to be pseudoautosomal regions (PAR), i.e., homologous sequences of nucleotides on the X and Y chromosomes. Therefore, this should be ok, the error is to put them in X or XY while the convention is to set them as XY ([link; section Sex-chromosome marker annotation](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5100670/))."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6983893e",
   "metadata": {},
   "source": [
    "We have also mitochondrial SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "6e196cb4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Index</th>\n",
       "      <th>Name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>GenTrain Score</th>\n",
       "      <th>SNP</th>\n",
       "      <th>ILMN Strand</th>\n",
       "      <th>Customer Strand</th>\n",
       "      <th>NormID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8158</th>\n",
       "      <td>8159</td>\n",
       "      <td>200610-1</td>\n",
       "      <td>MT</td>\n",
       "      <td>2755</td>\n",
       "      <td>0.6796</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>TOP</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8159</th>\n",
       "      <td>8160</td>\n",
       "      <td>200610-100</td>\n",
       "      <td>MT</td>\n",
       "      <td>15172</td>\n",
       "      <td>0.6310</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>TOP</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8160</th>\n",
       "      <td>8161</td>\n",
       "      <td>200610-102</td>\n",
       "      <td>MT</td>\n",
       "      <td>125</td>\n",
       "      <td>0.7739</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>BOT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8161</th>\n",
       "      <td>8162</td>\n",
       "      <td>200610-105</td>\n",
       "      <td>MT</td>\n",
       "      <td>236</td>\n",
       "      <td>0.6705</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>BOT</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8162</th>\n",
       "      <td>8163</td>\n",
       "      <td>200610-110</td>\n",
       "      <td>MT</td>\n",
       "      <td>3027</td>\n",
       "      <td>0.7420</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>BOT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545420</th>\n",
       "      <td>545421</td>\n",
       "      <td>rs62581340</td>\n",
       "      <td>MT</td>\n",
       "      <td>16356</td>\n",
       "      <td>0.6619</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>BOT</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545421</th>\n",
       "      <td>545422</td>\n",
       "      <td>rs62581341</td>\n",
       "      <td>MT</td>\n",
       "      <td>16362</td>\n",
       "      <td>0.6257</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>BOT</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>591382</th>\n",
       "      <td>591383</td>\n",
       "      <td>rs75214962</td>\n",
       "      <td>MT</td>\n",
       "      <td>11197</td>\n",
       "      <td>0.7111</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>TOP</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593585</th>\n",
       "      <td>593586</td>\n",
       "      <td>rs75577869</td>\n",
       "      <td>MT</td>\n",
       "      <td>13020</td>\n",
       "      <td>0.6882</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>TOP</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604840</th>\n",
       "      <td>604841</td>\n",
       "      <td>rs77417897</td>\n",
       "      <td>MT</td>\n",
       "      <td>13117</td>\n",
       "      <td>0.6551</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>TOP</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1138 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Index        Name Chromosome  Position  GenTrain Score    SNP  \\\n",
       "8158      8159    200610-1         MT      2755          0.6796  [T/C]   \n",
       "8159      8160  200610-100         MT     15172          0.6310  [T/C]   \n",
       "8160      8161  200610-102         MT       125          0.7739  [A/G]   \n",
       "8161      8162  200610-105         MT       236          0.6705  [A/G]   \n",
       "8162      8163  200610-110         MT      3027          0.7420  [A/G]   \n",
       "...        ...         ...        ...       ...             ...    ...   \n",
       "545420  545421  rs62581340         MT     16356          0.6619  [T/C]   \n",
       "545421  545422  rs62581341         MT     16362          0.6257  [T/C]   \n",
       "591382  591383  rs75214962         MT     11197          0.7111  [T/C]   \n",
       "593585  593586  rs75577869         MT     13020          0.6882  [T/C]   \n",
       "604840  604841  rs77417897         MT     13117          0.6551  [T/C]   \n",
       "\n",
       "       ILMN Strand Customer Strand  NormID  \n",
       "8158           BOT             TOP       6  \n",
       "8159           BOT             TOP       6  \n",
       "8160           TOP             BOT       1  \n",
       "8161           TOP             BOT       6  \n",
       "8162           TOP             BOT       1  \n",
       "...            ...             ...     ...  \n",
       "545420         BOT             BOT       3  \n",
       "545421         BOT             BOT     203  \n",
       "591382         BOT             TOP       3  \n",
       "593585         BOT             TOP       3  \n",
       "604840         BOT             TOP       4  \n",
       "\n",
       "[1138 rows x 9 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_map.loc[snp_map[\"Chromosome\"] == \"MT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "dca525cd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'The number of mito SNPs is less than 1500: True'"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'The number of mito SNPs is less than 1500: {snp_map.loc[snp_map[\"Chromosome\"] == \"MT\"].shape[0]<1500}'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c15e6c7b",
   "metadata": {},
   "source": [
    "Then, we have GenTrain score for each SNP."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "1d7cc89a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'There is no NA? True'"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'There is no NA? {all(snp_map[\"GenTrain Score\"].notna())}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "adb0e98d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Float? True'"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'Float? {snp_map[\"GenTrain Score\"].dtype == \"float64\"}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "44a90455",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count    654027.000000\n",
      "mean          0.807331\n",
      "std           0.087710\n",
      "min           0.000000\n",
      "25%           0.782200\n",
      "50%           0.817900\n",
      "75%           0.856800\n",
      "max           0.965100\n",
      "Name: GenTrain Score, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "print(snp_map[\"GenTrain Score\"].describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "884ef104",
   "metadata": {},
   "source": [
    "There is also allele and strand information. Not sure what strand is shown, but we are going to use foward in final report anyways."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b02bfd1d",
   "metadata": {},
   "source": [
    "The SNP I checked with mismatch between dbSNP and my database has also mismatch here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "b63ad351",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Index</th>\n",
       "      <th>Name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>GenTrain Score</th>\n",
       "      <th>SNP</th>\n",
       "      <th>ILMN Strand</th>\n",
       "      <th>Customer Strand</th>\n",
       "      <th>NormID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>652997</th>\n",
       "      <td>652998</td>\n",
       "      <td>rs999994</td>\n",
       "      <td>10</td>\n",
       "      <td>76975903</td>\n",
       "      <td>0.8048</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>TOP</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Index      Name Chromosome  Position  GenTrain Score    SNP  \\\n",
       "652997  652998  rs999994         10  76975903          0.8048  [A/G]   \n",
       "\n",
       "       ILMN Strand Customer Strand  NormID  \n",
       "652997         TOP             TOP       4  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_map.loc[snp_map[\"Name\"] == \"rs999994\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b877f20d",
   "metadata": {},
   "source": [
    "**Sample map**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2a19e9dc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Index</th>\n",
       "      <th>Name</th>\n",
       "      <th>ID</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Plate</th>\n",
       "      <th>Well</th>\n",
       "      <th>Group</th>\n",
       "      <th>Parent1</th>\n",
       "      <th>Parent2</th>\n",
       "      <th>Replicate</th>\n",
       "      <th>SentrixPosition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>205422690011_R01C01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7691CPSO</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>205422690011_R03C01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7688MRFO</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>205422690011_R05C01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7698ATMO</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>205422690011_R07C01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7695NKSO</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>205422690011_R09C01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6897TSSS</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>205422690123_R04C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>213</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6997EDFS</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>205422690123_R06C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6900CBNS</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>205422690123_R08C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>215</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6991HBJS</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>205422690123_R10C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>216</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1200ATFM</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>205422690123_R12C02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>216 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Index  Name        ID  Gender  Plate  Well  Group  Parent1  Parent2  \\\n",
       "0        1   NaN  7684BSAO    Male    NaN   NaN    NaN      NaN      NaN   \n",
       "1        2   NaN  7691CPSO    Male    NaN   NaN    NaN      NaN      NaN   \n",
       "2        3   NaN  7688MRFO    Male    NaN   NaN    NaN      NaN      NaN   \n",
       "3        4   NaN  7698ATMO  Female    NaN   NaN    NaN      NaN      NaN   \n",
       "4        5   NaN  7695NKSO    Male    NaN   NaN    NaN      NaN      NaN   \n",
       "..     ...   ...       ...     ...    ...   ...    ...      ...      ...   \n",
       "211    212   NaN  6897TSSS    Male    NaN   NaN    NaN      NaN      NaN   \n",
       "212    213   NaN  6997EDFS  Female    NaN   NaN    NaN      NaN      NaN   \n",
       "213    214   NaN  6900CBNS    Male    NaN   NaN    NaN      NaN      NaN   \n",
       "214    215   NaN  6991HBJS    Male    NaN   NaN    NaN      NaN      NaN   \n",
       "215    216   NaN  1200ATFM    Male    NaN   NaN    NaN      NaN      NaN   \n",
       "\n",
       "     Replicate      SentrixPosition  \n",
       "0          NaN  205422690011_R01C01  \n",
       "1          NaN  205422690011_R03C01  \n",
       "2          NaN  205422690011_R05C01  \n",
       "3          NaN  205422690011_R07C01  \n",
       "4          NaN  205422690011_R09C01  \n",
       "..         ...                  ...  \n",
       "211        NaN  205422690123_R04C02  \n",
       "212        NaN  205422690123_R06C02  \n",
       "213        NaN  205422690123_R08C02  \n",
       "214        NaN  205422690123_R10C02  \n",
       "215        NaN  205422690123_R12C02  \n",
       "\n",
       "[216 rows x 11 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_map = pd.read_csv(\"data/example_data/Sample_Map.txt\",\n",
    "                      delimiter=\"\\t\", \n",
    "                      header=0,\n",
    "                      low_memory=False) \n",
    "    #low_memory: Internally process the file in chunks, \n",
    "    #resulting in lower memory use while parsing, \n",
    "    #but possibly mixed type inference. To ensure no mixed \n",
    "    #types either set False, or specify the type with the dtype parameter. \n",
    "sample_map"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "673e2c5a",
   "metadata": {},
   "source": [
    "The number of samples should be 216 because this is the number of final reports present in the first batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "77b58359",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_map.shape[0] == 216"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e633c1e",
   "metadata": {},
   "source": [
    "We have index and ID for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "7b184177",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique cases are the number of rows? True\n",
      "Unique cases are the number of rows? True\n"
     ]
    }
   ],
   "source": [
    "print(f'Unique cases are the number of rows? {len(np.unique(sample_map[\"Index\"]))==sample_map.shape[0]}')\n",
    "print(f'Unique cases are the number of rows? {len(np.unique(sample_map[\"ID\"]))==sample_map.shape[0]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4d92bff",
   "metadata": {},
   "source": [
    "We have gender information, whith male, female and Unknown. This is estimated from illumina, not from the study!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "f8a36133",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Female', 'Male', 'Unknown'], dtype=object)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(sample_map[\"Gender\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "785f20bc",
   "metadata": {},
   "source": [
    "There are three individuals with unknown sex according to Illumina data. These are indeed the three problematic individuals according to the logR value, which is above the Illumina expecation of 0.3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "b3d8142f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Index</th>\n",
       "      <th>Name</th>\n",
       "      <th>ID</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Plate</th>\n",
       "      <th>Well</th>\n",
       "      <th>Group</th>\n",
       "      <th>Parent1</th>\n",
       "      <th>Parent2</th>\n",
       "      <th>Replicate</th>\n",
       "      <th>SentrixPosition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7699ZSSO</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>205422690078_R03C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>71</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3400SMJJ</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>205422690083_R10C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>150</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2498STAA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>205422690121_R11C01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Index  Name        ID   Gender  Plate  Well  Group  Parent1  Parent2  \\\n",
       "37      38   NaN  7699ZSSO  Unknown    NaN   NaN    NaN      NaN      NaN   \n",
       "70      71   NaN  3400SMJJ  Unknown    NaN   NaN    NaN      NaN      NaN   \n",
       "149    150   NaN  2498STAA  Unknown    NaN   NaN    NaN      NaN      NaN   \n",
       "\n",
       "     Replicate      SentrixPosition  \n",
       "37         NaN  205422690078_R03C02  \n",
       "70         NaN  205422690083_R10C02  \n",
       "149        NaN  205422690121_R11C01  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_map.loc[sample_map[\"Gender\"] == \"Unknown\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "7e52b4ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Unknown sex is below 10? True'"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'Unknown sex is below 10? {sample_map.loc[(sample_map[\"Gender\"] == \"Unknown\")].shape[0] < 10}'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06d2cab1",
   "metadata": {},
   "source": [
    "Then, we have columns for ID of the parents, but this is not the case in our study."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "713f57da",
   "metadata": {},
   "source": [
    "## Load phenotype data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1613e52",
   "metadata": {},
   "source": [
    "This include reported sex and VO2 max data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b2beaa1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gender</th>\n",
       "      <th>Age</th>\n",
       "      <th>Week 1 Body Mass</th>\n",
       "      <th>Week 8 Body Mass</th>\n",
       "      <th>Week 1 Beep test</th>\n",
       "      <th>Week 8 beep test</th>\n",
       "      <th>Week 1 Pred VO2max</th>\n",
       "      <th>Week 8 Pred VO2max</th>\n",
       "      <th>AGRF code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.0</td>\n",
       "      <td>9.6</td>\n",
       "      <td>10.3</td>\n",
       "      <td>45.2</td>\n",
       "      <td>47.7</td>\n",
       "      <td>8244TCOJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M</td>\n",
       "      <td>43.720548</td>\n",
       "      <td>79.7</td>\n",
       "      <td>80.8</td>\n",
       "      <td>8.6</td>\n",
       "      <td>9.1</td>\n",
       "      <td>41.8</td>\n",
       "      <td>43.6</td>\n",
       "      <td>6976MCJS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M</td>\n",
       "      <td>35.049315</td>\n",
       "      <td>86.2</td>\n",
       "      <td>86.2</td>\n",
       "      <td>10.2</td>\n",
       "      <td>10.7</td>\n",
       "      <td>47.4</td>\n",
       "      <td>49.0</td>\n",
       "      <td>3484JSJJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>F</td>\n",
       "      <td>23.561644</td>\n",
       "      <td>70.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>8.1</td>\n",
       "      <td>9.1</td>\n",
       "      <td>40.2</td>\n",
       "      <td>43.6</td>\n",
       "      <td>1389HSNM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M</td>\n",
       "      <td>26.164384</td>\n",
       "      <td>96.2</td>\n",
       "      <td>93.3</td>\n",
       "      <td>10.2</td>\n",
       "      <td>12.1</td>\n",
       "      <td>47.4</td>\n",
       "      <td>54.0</td>\n",
       "      <td>7593DBAO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1459</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2471AWMA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1460</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6897TSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1461</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6899DWOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1462</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6296TONJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1463</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7799BTDO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1464 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gender        Age  Week 1 Body Mass  Week 8 Body Mass  Week 1 Beep test  \\\n",
       "0         M  25.000000               NaN              75.0               9.6   \n",
       "1         M  43.720548              79.7              80.8               8.6   \n",
       "2         M  35.049315              86.2              86.2              10.2   \n",
       "3         F  23.561644              70.0              70.0               8.1   \n",
       "4         M  26.164384              96.2              93.3              10.2   \n",
       "...     ...        ...               ...               ...               ...   \n",
       "1459    NaN        NaN               NaN               NaN               NaN   \n",
       "1460    NaN        NaN               NaN               NaN               NaN   \n",
       "1461    NaN        NaN               NaN               NaN               NaN   \n",
       "1462    NaN        NaN               NaN               NaN               NaN   \n",
       "1463    NaN        NaN               NaN               NaN               NaN   \n",
       "\n",
       "     Week 8 beep test  Week 1 Pred VO2max  Week 8 Pred VO2max AGRF code  \n",
       "0                10.3                45.2                47.7  8244TCOJ  \n",
       "1                 9.1                41.8                43.6  6976MCJS  \n",
       "2                10.7                47.4                49.0  3484JSJJ  \n",
       "3                 9.1                40.2                43.6  1389HSNM  \n",
       "4                12.1                47.4                54.0  7593DBAO  \n",
       "...               ...                 ...                 ...       ...  \n",
       "1459              NaN                 NaN                 NaN  2471AWMA  \n",
       "1460              NaN                 NaN                 NaN  6897TSSS  \n",
       "1461              NaN                 NaN                 NaN  6899DWOS  \n",
       "1462              NaN                 NaN                 NaN  6296TONJ  \n",
       "1463              NaN                 NaN                 NaN  7799BTDO  \n",
       "\n",
       "[1464 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pheno_data = pd.read_csv(\"data/pheno_data/combact gene DNA GWAS 23062022_all_dna_samples.csv\",\n",
    "                      delimiter=\",\", \n",
    "                      header=0,\n",
    "                      low_memory=False) \n",
    "    #low_memory: Internally process the file in chunks, \n",
    "    #resulting in lower memory use while parsing, \n",
    "    #but possibly mixed type inference. To ensure no mixed \n",
    "    #types either set False, or specify the type with the dtype parameter. \n",
    "pheno_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b185656",
   "metadata": {},
   "source": [
    "We have gender (F/M) with some NAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6fc0b092",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count     1422\n",
       "unique       2\n",
       "top          M\n",
       "freq      1228\n",
       "Name: Gender, dtype: object"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pheno_data[\"Gender\"].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9af750da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NAs for sex: 42'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'NAs for sex: {pheno_data[pheno_data[\"Gender\"].isna()].shape[0]}'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3817e635",
   "metadata": {},
   "source": [
    "Also age, which is a float, as shown in the paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "41f34df6",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    1422.000000\n",
       "mean       21.969568\n",
       "std         5.065370\n",
       "min        17.000000\n",
       "25%        18.767808\n",
       "50%        20.163014\n",
       "75%        23.166438\n",
       "max        56.254237\n",
       "Name: Age, dtype: float64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pheno_data[\"Age\"].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcb74272",
   "metadata": {},
   "source": [
    "Then, different body mass and cardiorespiratory fitness variables.\n",
    "\n",
    "NOT CHECKED FOR NOW."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89b00150",
   "metadata": {},
   "source": [
    "We also have the ID of the sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9173bb3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       8244TCOJ\n",
       "1       6976MCJS\n",
       "2       3484JSJJ\n",
       "3       1389HSNM\n",
       "4       7593DBAO\n",
       "          ...   \n",
       "1459    2471AWMA\n",
       "1460    6897TSSS\n",
       "1461    6899DWOS\n",
       "1462    6296TONJ\n",
       "1463    7799BTDO\n",
       "Name: AGRF code, Length: 1464, dtype: object"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pheno_data[\"AGRF code\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "623e0a86",
   "metadata": {},
   "source": [
    "This code is the ID of the samples in the sample map of illumina.\n",
    "\n",
    "We have almost all samples of the first batch included in the pheno data, we only lack one individual."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "0894f86e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "215"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pheno_data[pheno_data[\"AGRF code\"].isin(sample_map[\"ID\"])].shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "070b0431",
   "metadata": {},
   "source": [
    "## Conversion of final report to Plink inputs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f490575",
   "metadata": {},
   "source": [
    "We are going to convert the illumina report (BPM file) to lgen file that can be used as input in Plink. This seems to be trivial ([link](https://www.biostars.org/p/51928/)), but we are going to use a tutorial just in case ([link](https://www.youtube.com/watch?v=5_pNby7l2dE&t=1s&ab_channel=GenomicsBootCamp), [link](https://pastebin.com/pzgw7JVp)).\n",
    "\n",
    "The final report has 1 line for the SNP of one individual. We have separated final reports for each individual. This is close to the lgen format of plink.\n",
    "\n",
    "Therefore, our goal is to create lgen, also the fam and map files required to load the lgen in Plink. Some information might be missing in the final report, so you need to replace them."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fb75964",
   "metadata": {},
   "source": [
    "**lgen file [plink info](https://www.cog-genomics.org/plink/1.9/formats#lgen)**\n",
    "\n",
    "A text file with no header line, and one line per genotype call (or just not-homozygous-major calls if 'lgen-ref' was invoked) usually with the following five fields:\n",
    "\n",
    "- Family ID\n",
    "- Within-family ID\n",
    "- Variant identifier\n",
    "- Allele call 1 ('0' for missing)\n",
    "- Allele call 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75d2fb72",
   "metadata": {},
   "source": [
    "As we have each sample in a separated final report, we need to bind them in order to get all the genotype calls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d1a759f",
   "metadata": {},
   "source": [
    "Get first the paths for each final report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "9f836a71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data/example_data/ILGSA24-17303_FinalReport1.txt',\n",
       " 'data/example_data/ILGSA24-17303_FinalReport2.txt']"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob\n",
    "list_reports_files_full_path = glob.glob(\"data/example_data/ILGSA24-17303_FinalReport*\") \n",
    "    #I prefer using the glob module, as it does pattern matching and expansion.\n",
    "        #https://stackoverflow.com/questions/3207219/how-do-i-list-all-files-of-a-directory\n",
    "list_reports_files_full_path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75071907",
   "metadata": {},
   "source": [
    "Load the data.frame selecting only the colums needed for the lgen file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "9cafb8c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "list_df_reports = [pd.read_csv(path,\n",
    "                               skiprows=10, #skip first 10 rows to get only the data\n",
    "                               usecols=[\"Sample ID\", #load only a subset of the columns\n",
    "                                        \"SNP Name\", \n",
    "                                        \"Allele1 - Forward\", \n",
    "                                        \"Allele2 - Forward\"],                               \n",
    "                               header=0,\n",
    "                               sep='\\t',\n",
    "                               low_memory=False) #to avoid problems with mixed types\n",
    "                   for path in list_reports_files_full_path]\n",
    "#for the 1000 files, we should parallelize"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d036eca",
   "metadata": {},
   "source": [
    "You can see how the selection of less columns decreases the size using the first final_report as example. We reduce the size almost 6 times."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "baa5753f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "631.959649\n",
      "163.38272\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "#get the size of final_report1 as a whole and then to only the selected columns\n",
    "#divide by 10^6 to get MBs\n",
    "print(sys.getsizeof(final_report1)/(10**6))\n",
    "print(sys.getsizeof(final_report1[[\"Sample ID\", \"SNP Name\", \"Allele1 - Forward\", \"Allele2 - Forward\"]])/(10**6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b261b907",
   "metadata": {},
   "source": [
    "The rest of necessary columns like sample ID or sample SNP and position/chromsome are just repeated across final_reports but in the sample and SNP map are not duplicated, so we can just use these smaller files to get these required columns. See next sections."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a2320c8",
   "metadata": {},
   "source": [
    "We get a list with one DF per final report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "b420a089",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>SNP Name</th>\n",
       "      <th>Allele1 - Forward</th>\n",
       "      <th>Allele2 - Forward</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:103380393</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:109439680</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:110198788</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:110201112</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:110201667</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654022</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>Y:8804321-A-C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654023</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>Y:8852748-A-C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654024</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>Y:8894070-A-T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654025</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>Y:9145267-T-C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654026</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>Y:9785736-A-G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>654027 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample ID       SNP Name Allele1 - Forward Allele2 - Forward\n",
       "0       7684BSAO    1:103380393                 A                 G\n",
       "1       7684BSAO    1:109439680                 A                 A\n",
       "2       7684BSAO    1:110198788                 T                 T\n",
       "3       7684BSAO    1:110201112                 C                 C\n",
       "4       7684BSAO    1:110201667                 C                 C\n",
       "...          ...            ...               ...               ...\n",
       "654022  7684BSAO  Y:8804321-A-C                 A                 A\n",
       "654023  7684BSAO  Y:8852748-A-C                 A                 A\n",
       "654024  7684BSAO  Y:8894070-A-T                 A                 A\n",
       "654025  7684BSAO  Y:9145267-T-C                 T                 T\n",
       "654026  7684BSAO  Y:9785736-A-G                 A                 A\n",
       "\n",
       "[654027 rows x 4 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_df_reports[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c18f6c3b",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Do we have as many DFs as paths we got? True'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'Do we have as many DFs as paths we got? {len(list_df_reports) == len(list_reports_files_full_path)}'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4f41eb0",
   "metadata": {},
   "source": [
    "Concatenate, because we have one row per SNP and sample, so we can just concateneate DFs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "8c49c355",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_final_report = pd.concat(objs=list_df_reports, #list DFs\n",
    "                              axis=0, #concatenate along rows\n",
    "                              ignore_index=True) #clear the existing index and reset it"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0de3b985",
   "metadata": {},
   "source": [
    "We should have as many rows as the total sum of rows across the list of DFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "29220156",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Do we have all the genotype calls? True'"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'Do we have all the genotype calls? {full_final_report.shape[0] == sum([element.shape[0] for element in list_df_reports])}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "a5e40979",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>SNP Name</th>\n",
       "      <th>Allele1 - Forward</th>\n",
       "      <th>Allele2 - Forward</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:103380393</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:109439680</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:110198788</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:110201112</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:110201667</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308049</th>\n",
       "      <td>7691CPSO</td>\n",
       "      <td>Y:8804321-A-C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308050</th>\n",
       "      <td>7691CPSO</td>\n",
       "      <td>Y:8852748-A-C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308051</th>\n",
       "      <td>7691CPSO</td>\n",
       "      <td>Y:8894070-A-T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308052</th>\n",
       "      <td>7691CPSO</td>\n",
       "      <td>Y:9145267-T-C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308053</th>\n",
       "      <td>7691CPSO</td>\n",
       "      <td>Y:9785736-A-G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1308054 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample ID       SNP Name Allele1 - Forward Allele2 - Forward\n",
       "0        7684BSAO    1:103380393                 A                 G\n",
       "1        7684BSAO    1:109439680                 A                 A\n",
       "2        7684BSAO    1:110198788                 T                 T\n",
       "3        7684BSAO    1:110201112                 C                 C\n",
       "4        7684BSAO    1:110201667                 C                 C\n",
       "...           ...            ...               ...               ...\n",
       "1308049  7691CPSO  Y:8804321-A-C                 A                 A\n",
       "1308050  7691CPSO  Y:8852748-A-C                 A                 A\n",
       "1308051  7691CPSO  Y:8894070-A-T                 A                 A\n",
       "1308052  7691CPSO  Y:9145267-T-C                 T                 T\n",
       "1308053  7691CPSO  Y:9785736-A-G                 A                 A\n",
       "\n",
       "[1308054 rows x 4 columns]"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_final_report"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d63bb2c",
   "metadata": {},
   "source": [
    "We have I and D as genotype calls, we will have to check that"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "008c4f39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['-', 'A', 'C', 'D', 'G', 'I', 'T'], dtype=object)"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(full_final_report[\"Allele1 - Forward\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "962228fd",
   "metadata": {},
   "source": [
    "Make a copy, using copy(), so modifications to the data or indices of the copy will not be reflected in the original object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "980b2525",
   "metadata": {},
   "outputs": [],
   "source": [
    "lgen_file_raw = full_final_report.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08ac90d4",
   "metadata": {},
   "source": [
    "Change to \"0\" those genotype calls with \"--\" to match the format of Plink"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "6f1bba37",
   "metadata": {},
   "outputs": [],
   "source": [
    "lgen_file_raw.loc[(lgen_file_raw[\"Allele1 - Forward\"] == \"-\") | \n",
    "                      (lgen_file_raw[\"Allele1 - Forward\"] == \"--\"), \n",
    "                      \"Allele1 - Forward\"] = \"0\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "b6a85053",
   "metadata": {},
   "outputs": [],
   "source": [
    "lgen_file_raw.loc[(lgen_file_raw[\"Allele2 - Forward\"] == \"-\") | \n",
    "                      (lgen_file_raw[\"Allele2 - Forward\"] == \"--\"), \n",
    "                      \"Allele2 - Forward\"] = \"0\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f41c52e4",
   "metadata": {},
   "source": [
    "Check we do not have \"--\" or \"-\" in alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "e41a445b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n",
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "print(\"-\" not in np.unique(lgen_raw[\"Allele1 - Forward\"]))\n",
    "print(\"--\" not in np.unique(lgen_raw[\"Allele1 - Forward\"]))\n",
    "print(\"-\" not in np.unique(lgen_raw[\"Allele2 - Forward\"]))\n",
    "print(\"--\" not in np.unique(lgen_raw[\"Allele2 - Forward\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0686c11",
   "metadata": {},
   "source": [
    "Add additional columns that are required for lgen files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "00d0e284",
   "metadata": {},
   "outputs": [],
   "source": [
    "lgen_file_raw[\"FID\"] = \"combat\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "be7d995b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>SNP Name</th>\n",
       "      <th>Allele1 - Forward</th>\n",
       "      <th>Allele2 - Forward</th>\n",
       "      <th>FID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:103380393</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>combat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:109439680</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>combat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:110198788</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>combat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:110201112</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>combat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:110201667</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>combat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308049</th>\n",
       "      <td>7691CPSO</td>\n",
       "      <td>Y:8804321-A-C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>combat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308050</th>\n",
       "      <td>7691CPSO</td>\n",
       "      <td>Y:8852748-A-C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>combat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308051</th>\n",
       "      <td>7691CPSO</td>\n",
       "      <td>Y:8894070-A-T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>combat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308052</th>\n",
       "      <td>7691CPSO</td>\n",
       "      <td>Y:9145267-T-C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>combat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308053</th>\n",
       "      <td>7691CPSO</td>\n",
       "      <td>Y:9785736-A-G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>combat</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1308054 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample ID       SNP Name Allele1 - Forward Allele2 - Forward     FID\n",
       "0        7684BSAO    1:103380393                 A                 G  combat\n",
       "1        7684BSAO    1:109439680                 A                 A  combat\n",
       "2        7684BSAO    1:110198788                 T                 T  combat\n",
       "3        7684BSAO    1:110201112                 C                 C  combat\n",
       "4        7684BSAO    1:110201667                 C                 C  combat\n",
       "...           ...            ...               ...               ...     ...\n",
       "1308049  7691CPSO  Y:8804321-A-C                 A                 A  combat\n",
       "1308050  7691CPSO  Y:8852748-A-C                 A                 A  combat\n",
       "1308051  7691CPSO  Y:8894070-A-T                 A                 A  combat\n",
       "1308052  7691CPSO  Y:9145267-T-C                 T                 T  combat\n",
       "1308053  7691CPSO  Y:9785736-A-G                 A                 A  combat\n",
       "\n",
       "[1308054 rows x 5 columns]"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lgen_file_raw"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b8636e1",
   "metadata": {},
   "source": [
    "Reorder the columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "e60011a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "lgen_file = lgen_file_raw[[\"FID\", \"Sample ID\", \"SNP Name\", \"Allele1 - Forward\", \"Allele2 - Forward\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "15be9c64",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>SNP Name</th>\n",
       "      <th>Allele1 - Forward</th>\n",
       "      <th>Allele2 - Forward</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>combat</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:103380393</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>combat</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:109439680</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>combat</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:110198788</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>combat</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:110201112</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>combat</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>1:110201667</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308049</th>\n",
       "      <td>combat</td>\n",
       "      <td>7691CPSO</td>\n",
       "      <td>Y:8804321-A-C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308050</th>\n",
       "      <td>combat</td>\n",
       "      <td>7691CPSO</td>\n",
       "      <td>Y:8852748-A-C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308051</th>\n",
       "      <td>combat</td>\n",
       "      <td>7691CPSO</td>\n",
       "      <td>Y:8894070-A-T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308052</th>\n",
       "      <td>combat</td>\n",
       "      <td>7691CPSO</td>\n",
       "      <td>Y:9145267-T-C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308053</th>\n",
       "      <td>combat</td>\n",
       "      <td>7691CPSO</td>\n",
       "      <td>Y:9785736-A-G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1308054 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            FID Sample ID       SNP Name Allele1 - Forward Allele2 - Forward\n",
       "0        combat  7684BSAO    1:103380393                 A                 G\n",
       "1        combat  7684BSAO    1:109439680                 A                 A\n",
       "2        combat  7684BSAO    1:110198788                 T                 T\n",
       "3        combat  7684BSAO    1:110201112                 C                 C\n",
       "4        combat  7684BSAO    1:110201667                 C                 C\n",
       "...         ...       ...            ...               ...               ...\n",
       "1308049  combat  7691CPSO  Y:8804321-A-C                 A                 A\n",
       "1308050  combat  7691CPSO  Y:8852748-A-C                 A                 A\n",
       "1308051  combat  7691CPSO  Y:8894070-A-T                 A                 A\n",
       "1308052  combat  7691CPSO  Y:9145267-T-C                 T                 T\n",
       "1308053  combat  7691CPSO  Y:9785736-A-G                 A                 A\n",
       "\n",
       "[1308054 rows x 5 columns]"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lgen_file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e2129c9",
   "metadata": {},
   "source": [
    "Save without header:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "4034eb0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "lgen_file.to_csv(\"data/plink_inputs_example/batch1_example.lgen\",\n",
    "                 sep=\"\\t\",\n",
    "                 header=None,\n",
    "                 index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "894f5463",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "**Fam file ([PLINK sample information file](https://www.cog-genomics.org/plink/1.9/formats#fam))**\n",
    "\n",
    "A text file with no header line, and one line per sample with the following six fields:\n",
    "\n",
    "- Family ID ('FID')\n",
    "- Within-family ID ('IID'; cannot be '0')\n",
    "- Within-family ID of father ('0' if father isn't in dataset)\n",
    "- Within-family ID of mother ('0' if mother isn't in dataset)\n",
    "- Sex code ('1' = male, '2' = female, '0' = unknown)\n",
    "- Phenotype value ('1' = control, '2' = case, '-9'/'0'/non-numeric = missing data if case/control)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89ab00df",
   "metadata": {},
   "source": [
    "We can get the self-reported sex and sample ID from the pheno data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "c5061bbf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gender</th>\n",
       "      <th>AGRF code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M</td>\n",
       "      <td>8244TCOJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M</td>\n",
       "      <td>6976MCJS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M</td>\n",
       "      <td>3484JSJJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>F</td>\n",
       "      <td>1389HSNM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M</td>\n",
       "      <td>7593DBAO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1459</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2471AWMA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1460</th>\n",
       "      <td>NaN</td>\n",
       "      <td>6897TSSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1461</th>\n",
       "      <td>NaN</td>\n",
       "      <td>6899DWOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1462</th>\n",
       "      <td>NaN</td>\n",
       "      <td>6296TONJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1463</th>\n",
       "      <td>NaN</td>\n",
       "      <td>7799BTDO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1464 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gender AGRF code\n",
       "0         M  8244TCOJ\n",
       "1         M  6976MCJS\n",
       "2         M  3484JSJJ\n",
       "3         F  1389HSNM\n",
       "4         M  7593DBAO\n",
       "...     ...       ...\n",
       "1459    NaN  2471AWMA\n",
       "1460    NaN  6897TSSS\n",
       "1461    NaN  6899DWOS\n",
       "1462    NaN  6296TONJ\n",
       "1463    NaN  7799BTDO\n",
       "\n",
       "[1464 rows x 2 columns]"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fam_file_raw = pheno_data.loc[:, [\"Gender\", \"AGRF code\"]]\n",
    "fam_file_raw"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4347b899",
   "metadata": {},
   "source": [
    "Codify the sex variable following plink notation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "6fd520c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['0', '1', '2'], dtype=object)"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fam_file_raw.loc[fam_file_raw[\"Gender\"].isna(), \"Gender\"] = \"0\"\n",
    "fam_file_raw.loc[fam_file_raw[\"Gender\"] == \"M\", \"Gender\"] = \"1\"\n",
    "fam_file_raw.loc[fam_file_raw[\"Gender\"] == \"F\", \"Gender\"] = \"2\"\n",
    "np.unique(fam_file_raw[\"Gender\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8d5a3ec",
   "metadata": {},
   "source": [
    "Add the family variables and the phenotype (not added for now)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "ec696bf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "fam_file_raw[\"FID\"] = \"combat\" #ID for the whole study\n",
    "fam_file_raw[\"IID_father\"] = \"0\"\n",
    "fam_file_raw[\"IID_mother\"] = \"0\"\n",
    "fam_file_raw[\"phenotype\"] = -9 #this is no data for phenotype"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f88a31a9",
   "metadata": {},
   "source": [
    "Reorder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "62a874e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "fam_file = fam_file_raw[[\"FID\", \"AGRF code\", \"IID_father\", \"IID_mother\", \"Gender\", \"phenotype\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "b78872c2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>AGRF code</th>\n",
       "      <th>IID_father</th>\n",
       "      <th>IID_mother</th>\n",
       "      <th>Gender</th>\n",
       "      <th>phenotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>combat</td>\n",
       "      <td>8244TCOJ</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>combat</td>\n",
       "      <td>6976MCJS</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>combat</td>\n",
       "      <td>3484JSJJ</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>combat</td>\n",
       "      <td>1389HSNM</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>combat</td>\n",
       "      <td>7593DBAO</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1459</th>\n",
       "      <td>combat</td>\n",
       "      <td>2471AWMA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1460</th>\n",
       "      <td>combat</td>\n",
       "      <td>6897TSSS</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1461</th>\n",
       "      <td>combat</td>\n",
       "      <td>6899DWOS</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1462</th>\n",
       "      <td>combat</td>\n",
       "      <td>6296TONJ</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1463</th>\n",
       "      <td>combat</td>\n",
       "      <td>7799BTDO</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1464 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         FID AGRF code IID_father IID_mother Gender  phenotype\n",
       "0     combat  8244TCOJ          0          0      1         -9\n",
       "1     combat  6976MCJS          0          0      1         -9\n",
       "2     combat  3484JSJJ          0          0      1         -9\n",
       "3     combat  1389HSNM          0          0      2         -9\n",
       "4     combat  7593DBAO          0          0      1         -9\n",
       "...      ...       ...        ...        ...    ...        ...\n",
       "1459  combat  2471AWMA          0          0      0         -9\n",
       "1460  combat  6897TSSS          0          0      0         -9\n",
       "1461  combat  6899DWOS          0          0      0         -9\n",
       "1462  combat  6296TONJ          0          0      0         -9\n",
       "1463  combat  7799BTDO          0          0      0         -9\n",
       "\n",
       "[1464 rows x 6 columns]"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fam_file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45abb817",
   "metadata": {},
   "source": [
    "Save without header:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "cd829b27",
   "metadata": {},
   "outputs": [],
   "source": [
    "fam_file.to_csv(\"data/plink_inputs_example/batch1_example.fam\",\n",
    "                sep=\"\\t\",\n",
    "                header=None,\n",
    "                index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "579587a7",
   "metadata": {},
   "source": [
    "**Map file [Plink info](https://www.cog-genomics.org/plink/1.9/formats#map)**\n",
    "\n",
    "A text file with no header line, and one line per variant with the following 3-4 fields:\n",
    "\n",
    "- Chromosome code. PLINK 1.9 also permits contig names here, but most older programs do not.\n",
    "- Variant identifier\n",
    "- Position in morgans or centimorgans (optional; also safe to use dummy value of '0')\n",
    "- Base-pair coordinate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fed1c05",
   "metadata": {},
   "source": [
    "We can get this information from the SNP map.\n",
    "\n",
    "We could also get this information from the final report but, as shown in previous lines, we have only loaded the columns required for the lgen file because we have to load hundreds of final reports and concatenate them, so it will use less memory with less columns. The Indexes and positions are not duplicated in SNP map, only 1 row per SNP, being a much smaller file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "48e63632",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Index</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>SNP Index</th>\n",
       "      <th>SNP Name</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Position</th>\n",
       "      <th>GT Score</th>\n",
       "      <th>GC Score</th>\n",
       "      <th>Allele1 - AB</th>\n",
       "      <th>...</th>\n",
       "      <th>Y Raw</th>\n",
       "      <th>X</th>\n",
       "      <th>Y</th>\n",
       "      <th>B Allele Freq</th>\n",
       "      <th>Log R Ratio</th>\n",
       "      <th>SNP Aux</th>\n",
       "      <th>SNP</th>\n",
       "      <th>ILMN Strand</th>\n",
       "      <th>Top Genomic Sequence</th>\n",
       "      <th>Customer Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1:103380393</td>\n",
       "      <td>1</td>\n",
       "      <td>102914837</td>\n",
       "      <td>0.7987</td>\n",
       "      <td>0.8136</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>958</td>\n",
       "      <td>0.324</td>\n",
       "      <td>0.360</td>\n",
       "      <td>0.4849</td>\n",
       "      <td>0.3097</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1:109439680</td>\n",
       "      <td>1</td>\n",
       "      <td>108897058</td>\n",
       "      <td>0.8792</td>\n",
       "      <td>0.4803</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>131</td>\n",
       "      <td>0.779</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2660</td>\n",
       "      <td>0</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>1:110198788</td>\n",
       "      <td>1</td>\n",
       "      <td>109656166</td>\n",
       "      <td>0.8780</td>\n",
       "      <td>0.9154</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>198</td>\n",
       "      <td>1.170</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0312</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>1:110201112</td>\n",
       "      <td>1</td>\n",
       "      <td>109658490</td>\n",
       "      <td>0.8584</td>\n",
       "      <td>0.4545</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>728</td>\n",
       "      <td>1.154</td>\n",
       "      <td>0.072</td>\n",
       "      <td>0.0188</td>\n",
       "      <td>-0.0746</td>\n",
       "      <td>0</td>\n",
       "      <td>[G/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>1:110201667</td>\n",
       "      <td>1</td>\n",
       "      <td>109659045</td>\n",
       "      <td>0.7815</td>\n",
       "      <td>0.7185</td>\n",
       "      <td>B</td>\n",
       "      <td>...</td>\n",
       "      <td>3269</td>\n",
       "      <td>0.120</td>\n",
       "      <td>1.291</td>\n",
       "      <td>0.9797</td>\n",
       "      <td>-0.0749</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654022</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>654023</td>\n",
       "      <td>Y:8804321-A-C</td>\n",
       "      <td>Y</td>\n",
       "      <td>8936280</td>\n",
       "      <td>0.7229</td>\n",
       "      <td>0.4350</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>157</td>\n",
       "      <td>0.403</td>\n",
       "      <td>0.039</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-1.2565</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/G]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654023</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>654024</td>\n",
       "      <td>Y:8852748-A-C</td>\n",
       "      <td>Y</td>\n",
       "      <td>8984707</td>\n",
       "      <td>0.8575</td>\n",
       "      <td>0.4533</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>147</td>\n",
       "      <td>0.730</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.3919</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/G]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654024</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>654025</td>\n",
       "      <td>Y:8894070-A-T</td>\n",
       "      <td>Y</td>\n",
       "      <td>9026029</td>\n",
       "      <td>0.8175</td>\n",
       "      <td>0.4061</td>\n",
       "      <td>B</td>\n",
       "      <td>...</td>\n",
       "      <td>3099</td>\n",
       "      <td>0.017</td>\n",
       "      <td>1.436</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0729</td>\n",
       "      <td>0</td>\n",
       "      <td>[A/T]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654025</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>654026</td>\n",
       "      <td>Y:9145267-T-C</td>\n",
       "      <td>Y</td>\n",
       "      <td>9307658</td>\n",
       "      <td>0.7893</td>\n",
       "      <td>0.2966</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>201</td>\n",
       "      <td>0.334</td>\n",
       "      <td>0.059</td>\n",
       "      <td>0.0178</td>\n",
       "      <td>-1.3112</td>\n",
       "      <td>0</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654026</th>\n",
       "      <td>1</td>\n",
       "      <td>7684BSAO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>654027</td>\n",
       "      <td>Y:9785736-A-G</td>\n",
       "      <td>Y</td>\n",
       "      <td>9948127</td>\n",
       "      <td>0.8452</td>\n",
       "      <td>0.2421</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>124</td>\n",
       "      <td>0.260</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-1.7377</td>\n",
       "      <td>0</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>654027 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample Index Sample ID  Sample Name  SNP Index       SNP Name Chr  \\\n",
       "0                  1  7684BSAO          NaN          1    1:103380393   1   \n",
       "1                  1  7684BSAO          NaN          2    1:109439680   1   \n",
       "2                  1  7684BSAO          NaN          3    1:110198788   1   \n",
       "3                  1  7684BSAO          NaN          4    1:110201112   1   \n",
       "4                  1  7684BSAO          NaN          5    1:110201667   1   \n",
       "...              ...       ...          ...        ...            ...  ..   \n",
       "654022             1  7684BSAO          NaN     654023  Y:8804321-A-C   Y   \n",
       "654023             1  7684BSAO          NaN     654024  Y:8852748-A-C   Y   \n",
       "654024             1  7684BSAO          NaN     654025  Y:8894070-A-T   Y   \n",
       "654025             1  7684BSAO          NaN     654026  Y:9145267-T-C   Y   \n",
       "654026             1  7684BSAO          NaN     654027  Y:9785736-A-G   Y   \n",
       "\n",
       "         Position  GT Score  GC Score Allele1 - AB  ... Y Raw      X      Y  \\\n",
       "0       102914837    0.7987    0.8136            A  ...   958  0.324  0.360   \n",
       "1       108897058    0.8792    0.4803            A  ...   131  0.779  0.019   \n",
       "2       109656166    0.8780    0.9154            A  ...   198  1.170  0.028   \n",
       "3       109658490    0.8584    0.4545            A  ...   728  1.154  0.072   \n",
       "4       109659045    0.7815    0.7185            B  ...  3269  0.120  1.291   \n",
       "...           ...       ...       ...          ...  ...   ...    ...    ...   \n",
       "654022    8936280    0.7229    0.4350            A  ...   157  0.403  0.039   \n",
       "654023    8984707    0.8575    0.4533            A  ...   147  0.730  0.028   \n",
       "654024    9026029    0.8175    0.4061            B  ...  3099  0.017  1.436   \n",
       "654025    9307658    0.7893    0.2966            A  ...   201  0.334  0.059   \n",
       "654026    9948127    0.8452    0.2421            A  ...   124  0.260  0.029   \n",
       "\n",
       "       B Allele Freq Log R Ratio SNP Aux    SNP  ILMN Strand  \\\n",
       "0             0.4849      0.3097       0  [T/C]          BOT   \n",
       "1             0.0000      0.2660       0  [A/G]          TOP   \n",
       "2             0.0000      0.0312       0  [T/C]          BOT   \n",
       "3             0.0188     -0.0746       0  [G/C]          BOT   \n",
       "4             0.9797     -0.0749       0  [T/C]          BOT   \n",
       "...              ...         ...     ...    ...          ...   \n",
       "654022        0.0000     -1.2565       0  [T/G]          BOT   \n",
       "654023        0.0000     -0.3919       0  [T/G]          BOT   \n",
       "654024        1.0000      0.0729       0  [A/T]          TOP   \n",
       "654025        0.0178     -1.3112       0  [T/C]          BOT   \n",
       "654026        0.0000     -1.7377       0  [A/G]          TOP   \n",
       "\n",
       "        Top Genomic Sequence  Customer Strand  \n",
       "0                        NaN              TOP  \n",
       "1                        NaN              TOP  \n",
       "2                        NaN              BOT  \n",
       "3                        NaN              TOP  \n",
       "4                        NaN              BOT  \n",
       "...                      ...              ...  \n",
       "654022                   NaN              TOP  \n",
       "654023                   NaN              TOP  \n",
       "654024                   NaN              BOT  \n",
       "654025                   NaN              BOT  \n",
       "654026                   NaN              TOP  \n",
       "\n",
       "[654027 rows x 30 columns]"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_report1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "135b8e54",
   "metadata": {},
   "source": [
    "Make a copy with copy(), so modifications to the data or indices of the copy will not be reflected in the original object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "97569848",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Index</th>\n",
       "      <th>Name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>GenTrain Score</th>\n",
       "      <th>SNP</th>\n",
       "      <th>ILMN Strand</th>\n",
       "      <th>Customer Strand</th>\n",
       "      <th>NormID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1:103380393</td>\n",
       "      <td>1</td>\n",
       "      <td>102914837</td>\n",
       "      <td>0.7987</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>TOP</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1:109439680</td>\n",
       "      <td>1</td>\n",
       "      <td>108897058</td>\n",
       "      <td>0.8792</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>TOP</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1:110198788</td>\n",
       "      <td>1</td>\n",
       "      <td>109656166</td>\n",
       "      <td>0.8780</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>BOT</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>1:110201112</td>\n",
       "      <td>1</td>\n",
       "      <td>109658490</td>\n",
       "      <td>0.8584</td>\n",
       "      <td>[G/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>TOP</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1:110201667</td>\n",
       "      <td>1</td>\n",
       "      <td>109659045</td>\n",
       "      <td>0.7815</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>BOT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654022</th>\n",
       "      <td>654023</td>\n",
       "      <td>Y:8804321-A-C</td>\n",
       "      <td>Y</td>\n",
       "      <td>8936280</td>\n",
       "      <td>0.7229</td>\n",
       "      <td>[T/G]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>TOP</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654023</th>\n",
       "      <td>654024</td>\n",
       "      <td>Y:8852748-A-C</td>\n",
       "      <td>Y</td>\n",
       "      <td>8984707</td>\n",
       "      <td>0.8575</td>\n",
       "      <td>[T/G]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>TOP</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654024</th>\n",
       "      <td>654025</td>\n",
       "      <td>Y:8894070-A-T</td>\n",
       "      <td>Y</td>\n",
       "      <td>9026029</td>\n",
       "      <td>0.8175</td>\n",
       "      <td>[A/T]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>BOT</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654025</th>\n",
       "      <td>654026</td>\n",
       "      <td>Y:9145267-T-C</td>\n",
       "      <td>Y</td>\n",
       "      <td>9307658</td>\n",
       "      <td>0.7893</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>BOT</td>\n",
       "      <td>BOT</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654026</th>\n",
       "      <td>654027</td>\n",
       "      <td>Y:9785736-A-G</td>\n",
       "      <td>Y</td>\n",
       "      <td>9948127</td>\n",
       "      <td>0.8452</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>TOP</td>\n",
       "      <td>TOP</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>654027 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Index           Name Chromosome   Position  GenTrain Score    SNP  \\\n",
       "0            1    1:103380393          1  102914837          0.7987  [T/C]   \n",
       "1            2    1:109439680          1  108897058          0.8792  [A/G]   \n",
       "2            3    1:110198788          1  109656166          0.8780  [T/C]   \n",
       "3            4    1:110201112          1  109658490          0.8584  [G/C]   \n",
       "4            5    1:110201667          1  109659045          0.7815  [T/C]   \n",
       "...        ...            ...        ...        ...             ...    ...   \n",
       "654022  654023  Y:8804321-A-C          Y    8936280          0.7229  [T/G]   \n",
       "654023  654024  Y:8852748-A-C          Y    8984707          0.8575  [T/G]   \n",
       "654024  654025  Y:8894070-A-T          Y    9026029          0.8175  [A/T]   \n",
       "654025  654026  Y:9145267-T-C          Y    9307658          0.7893  [T/C]   \n",
       "654026  654027  Y:9785736-A-G          Y    9948127          0.8452  [A/G]   \n",
       "\n",
       "       ILMN Strand Customer Strand  NormID  \n",
       "0              BOT             TOP      28  \n",
       "1              TOP             TOP       1  \n",
       "2              BOT             BOT       4  \n",
       "3              BOT             TOP     101  \n",
       "4              BOT             BOT       1  \n",
       "...            ...             ...     ...  \n",
       "654022         BOT             TOP       3  \n",
       "654023         BOT             TOP       3  \n",
       "654024         TOP             BOT     203  \n",
       "654025         BOT             BOT       3  \n",
       "654026         TOP             TOP       3  \n",
       "\n",
       "[654027 rows x 9 columns]"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "map_file_raw = snp_map.copy()\n",
    "map_file_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "7d0c338a",
   "metadata": {},
   "outputs": [],
   "source": [
    "map_file_raw[\"centimorgans\"] = 0 #dummy value of zero"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "6ba7f864",
   "metadata": {},
   "outputs": [],
   "source": [
    "map_file = map_file_raw[[\"Chromosome\", \"Name\", \"centimorgans\", \"Position\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "b4100062",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Name</th>\n",
       "      <th>centimorgans</th>\n",
       "      <th>Position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1:103380393</td>\n",
       "      <td>0</td>\n",
       "      <td>102914837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1:109439680</td>\n",
       "      <td>0</td>\n",
       "      <td>108897058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1:110198788</td>\n",
       "      <td>0</td>\n",
       "      <td>109656166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1:110201112</td>\n",
       "      <td>0</td>\n",
       "      <td>109658490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1:110201667</td>\n",
       "      <td>0</td>\n",
       "      <td>109659045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654022</th>\n",
       "      <td>Y</td>\n",
       "      <td>Y:8804321-A-C</td>\n",
       "      <td>0</td>\n",
       "      <td>8936280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654023</th>\n",
       "      <td>Y</td>\n",
       "      <td>Y:8852748-A-C</td>\n",
       "      <td>0</td>\n",
       "      <td>8984707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654024</th>\n",
       "      <td>Y</td>\n",
       "      <td>Y:8894070-A-T</td>\n",
       "      <td>0</td>\n",
       "      <td>9026029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654025</th>\n",
       "      <td>Y</td>\n",
       "      <td>Y:9145267-T-C</td>\n",
       "      <td>0</td>\n",
       "      <td>9307658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654026</th>\n",
       "      <td>Y</td>\n",
       "      <td>Y:9785736-A-G</td>\n",
       "      <td>0</td>\n",
       "      <td>9948127</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>654027 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome           Name  centimorgans   Position\n",
       "0               1    1:103380393             0  102914837\n",
       "1               1    1:109439680             0  108897058\n",
       "2               1    1:110198788             0  109656166\n",
       "3               1    1:110201112             0  109658490\n",
       "4               1    1:110201667             0  109659045\n",
       "...           ...            ...           ...        ...\n",
       "654022          Y  Y:8804321-A-C             0    8936280\n",
       "654023          Y  Y:8852748-A-C             0    8984707\n",
       "654024          Y  Y:8894070-A-T             0    9026029\n",
       "654025          Y  Y:9145267-T-C             0    9307658\n",
       "654026          Y  Y:9785736-A-G             0    9948127\n",
       "\n",
       "[654027 rows x 4 columns]"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "map_file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99ffb579",
   "metadata": {},
   "source": [
    "Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "524bc7ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "map_file.to_csv(\"data/plink_inputs_example/batch1_example.map\",\n",
    "                sep=\"\\t\",\n",
    "                header=None,\n",
    "                index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8313466e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0025be6c",
   "metadata": {},
   "source": [
    "CHECK THE THREE FILES"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f1bde0e",
   "metadata": {},
   "source": [
    "Remove SNPs/samples not included in the lgen file!!! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c4da7ec",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4a6d26b6",
   "metadata": {},
   "source": [
    "Change to ped file with PLINK "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "03777d09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "batch1_example.fam\n",
      "batch1_example.lgen\n",
      "batch1_example.map\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os \n",
    "\n",
    "os.system(\"ls data/plink_inputs_example\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "97dde29e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b6.27 64-bit (10 Dec 2022)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2022 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to batch1_example_plink.log.\n",
      "Options in effect:\n",
      "  --lfile batch1_example\n",
      "  --nonfounders\n",
      "  --out batch1_example_plink\n",
      "  --recode\n",
      "\n",
      "64162 MB RAM detected; reserving 32081 MB for main workspace.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: Duplicate ID 'combat 1100JHJM'.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "768"
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os \n",
    "\n",
    "os.system(\"cd data/plink_inputs_example; plink --nonfounders --lfile batch1_example --recode --out batch1_example_plink\")\n",
    "\n",
    "#I have not checked this code"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5362ee4",
   "metadata": {},
   "source": [
    "This error is caused because in the phenotype file, row 1423 is empty and then we have rows with sample ID but no phenotype, so we should remove all rows from 1423, "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "460460bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "fam_file_raw.iloc[1422,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27d51676",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "080cb124",
   "metadata": {},
   "source": [
    "Full tutorial for plink\n",
    "\n",
    "https://genomicsbootcamp.github.io/book/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11a7fbfb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7138f635",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8d418685",
   "metadata": {},
   "source": [
    "Things to check when QC:\n",
    "\n",
    "- Check the genome build. I have manually checked some SNPs and they have the position of hg38.\n",
    "    - YOU ARE NOT ADDED CENTIMORGANS, SO BE CAREFUL WITH WHAT PLINK DO ABOUT LD.\n",
    "- Remove zero chromosomes\n",
    "    - also MT, X, Y and XY?\n",
    "- Check strand, because the foward strand of illumina not always matches that of dbSNP\n",
    "    - StrandScript is a software to check and solve that\n",
    "- Check the illumina pdf report for the first batch (zip file in initial stuff), because they say there are three problematic samples.\n",
    "    - these three sample also have unknown sex according to illumina data!!\n",
    "- check sex between illumina and real sex data (pheno_data)\n",
    "    - remove uknown sex individuals?\n",
    "- check genotype calls that are I or D\n",
    "    - how plink deals with this?\n",
    "- ask David about the samples without phenotype in the excel file?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf04506b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b8acf1b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de52985b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "cd /home/dftortosa/Desktop\n",
    "\n",
    "plink --file ILGSA24-17873 --freq --out ILGSA24-17873_analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "711f6de4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a108c8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "cd data/example_data\n",
    "\n",
    "#https://www.cog-genomics.org/plink/1.9/formats#lgen\n",
    "\n",
    "#plink --lfile example_ILGSA24-17303_FinalReport1.txt --recode\n",
    "\n",
    "#https://www.biostars.org/p/51928/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02e72dad",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc1ecae0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2dde7afe",
   "metadata": {},
   "source": [
    "There are two batches (ILGSA24-17303 and ILGSA24-17873), being the data separated for these. \n",
    "\n",
    "- In ILGSA24-17303.zip, we have the final reports for each 216 samples, along with the sample and snp maps.\n",
    "    - In the initial_stuff folder there is a zip called \"ILGSA24-17303.zip\" that I may downloaded from the initial location where this data was stored in summer 2022. There are Plink files, but I am not sure this is the correct data and I cannot find the final_report files.\n",
    "\n",
    "- In 17873, we have the IDAT files with probs intensity from the microarrays used to genotype (first zips), the final reports (CAGRF20093767.zip) and a inputs for plink. But all of this only for 1248 individuals, not the whole cohort.\n",
    "    - CAGRF20093767.zip includes the final reports of 1248 individuals, along with the sample and snp maps."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b06367e",
   "metadata": {},
   "source": [
    "warning [CAGRF20093767.zip]:  32332459452 extra bytes at beginning or within zipfile\n",
    "  (attempting to process anyway)\n",
    "error [CAGRF20093767.zip]:  start of central directory not found;\n",
    "  zipfile corrupt.\n",
    "  (please check that you have transferred or created the zipfile in the\n",
    "  appropriate BINARY mode and that you have compiled UnZip properly)\n",
    "\n",
    "\n",
    "CHECK WARNING\n",
    "\n",
    "solution for the error: https://askubuntu.com/questions/54904/unzip-error-end-of-central-directory-signature-not-found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dcfa8f1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ebe5e19",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "842b45b9",
   "metadata": {},
   "source": [
    "- Quality Control Procedures for Genome-Wide Association Studies\n",
    "- A tutorial on conducting genome‐wide association studies: Quality control and statistical analysis\n",
    "- Genetic prediction of complex traits with polygenic scores: a statistical review\n",
    "- Addressing the challenges of polygenic scores in human genetic research"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
